<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
  <head>
    <title>20191009_tidyr.utf8</title>
    <meta charset="utf-8" />
    <meta name="author" content="Rockefeller University, Bioinformatics Resource Centre" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="metropolisCustom.css" type="text/css" />
    <link rel="stylesheet" href="metropolis-fontsCustom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Data wrangling with tidy
<html>
<div style="float:left">

</div>
<hr color='#EB811B' size=1px width=796px>
</html>
### Rockefeller University, Bioinformatics Resource Centre
### <a href="https://rockefelleruniversity.github.io/RU_tidyverse/" class="uri">https://rockefelleruniversity.github.io/RU_tidyverse/</a>

---




---
# Data wrangling with Tidy

&lt;b&gt;Some problems we face when dealing with data regularly.&lt;/b&gt;

&lt;ul&gt;
  &lt;li&gt;Every dataset is different. Sometimes very different.&lt;/li&gt;
  &lt;li&gt;There are many ways to do things. Everyone has their favorite syntax.&lt;/li&gt;
&lt;/ul&gt;

&lt;b&gt;The issue:&lt;/b&gt;
Many fundamental data processing functions exist in &lt;i&gt;Base R&lt;/i&gt; and beyond. Sometimes they can be inconsistent or unnecessarily complex. Especially when dealing with non-standard dataframes. The result is code that is confusing and doesn't flow i.e. nested functions

---
## What does it mean to be tidy?

Tidyverse is most importantly a philosophy for data analysis that more often then not makes wrangling data easier. The tidyverse community have built what they describe as an &lt;i&gt;opinionated&lt;/i&gt; group of packages. These packages readily talk to one another. 

&lt;ul&gt;
  &lt;li&gt;More efficient code&lt;/li&gt;
  &lt;li&gt;Easier to remember syntax&lt;/li&gt;
  &lt;li&gt;Easier to read syntax&lt;/li&gt;
&lt;/ul&gt;

You can read their &lt;a href="https://cran.r-project.org/web/packages/tidyverse/vignettes/manifesto.html"&gt;manifesto&lt;/a&gt; to get a better understanding of the tidy ethos. 
  
---
## What does it &lt;i&gt;actually&lt;/i&gt; mean to be tidy?

&lt;ul&gt;
  &lt;li&gt;A defined vision for coding style in R  
  &lt;li&gt;A defined vision for data formats in R
  &lt;li&gt;A defined vision for package design in R
  &lt;li&gt;Unified set of community pushing in a cohesive direction
  &lt;li&gt;Critical mass of people to influence the way the whole R community evolves
&lt;/ul&gt;

---
## What are the main tidy tools?

&lt;ul&gt;
  &lt;li&gt;ggplot2 – making pretty graphs 
  &lt;li&gt;readr – reading data into R  
  &lt;li&gt;dplyr – manipulating data  
  &lt;li&gt;tibble - working with tibbles  
  &lt;li&gt;tidyr – miscellaneous tools for tidying data
  &lt;li&gt;purrr - iterating over data  
  &lt;li&gt;stringr – working with strings  
  &lt;li&gt;forcats - working with factors  
&lt;/ul&gt;

Other tools have now been made for the tidy community. This community also overlaps with bioconductor. But the packages above are the linchpins that hold it together. 

---
## What are we doing today

Workflow Image for working with data.

---
## What we won't be doing today

We already have a course online for plotting, including ggplot.     

https://rockefelleruniversity.github.io/Plotting_In_R/

---
# Lets get tidy!
First step lets load in the data we are using today

```r
cat(getwd())
```

```
## /Users/mattpaul/Documents/Box Sync/RU/Teaching/RU_side/RU_tidyverse/tidyR
```

```r
load(file='dataset/my_tidy.Rdata')
```

---
## Are all data frames equal?

```r
head(df1)
```

```
## # A tibble: 6 x 5
##   salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml
##       &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
## 1     35032 Chinook salmon yearling                 147       41.3
## 2     35035 Sockeye salmon juvenile                 121       NA  
## 3     35036 Sockeye salmon juvenile                 112       NA  
## 4     35037 Steelhead      juvenile                 220       42.7
## 5     35038 Steelhead      juvenile                 152       NA  
## 6     35033 Chinook salmon mixed age juvenile       444       62.1
```


```r
head(df2)
```

```
## # A tibble: 6 x 5
##   salmon_id common_name    age_classbylength  variable   value
##       &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;              &lt;fct&gt;      &lt;dbl&gt;
## 1     35032 Chinook salmon yearling           length_mm  147  
## 2     35032 Chinook salmon yearling           IGF1_ng_ml  41.3
## 3     35033 Chinook salmon mixed age juvenile length_mm  444  
## 4     35033 Chinook salmon mixed age juvenile IGF1_ng_ml  62.1
## 5     35034 Sockeye salmon juvenile           length_mm  139  
## 6     35034 Sockeye salmon juvenile           IGF1_ng_ml  NA
```

---
### Are all data frames equal?

```r
head(df3a)
```

```
## # A tibble: 6 x 5
##   salmon_id common_name    age_classbylength  variable  value
##       &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;              &lt;fct&gt;     &lt;dbl&gt;
## 1     35032 Chinook salmon yearling           length_mm   147
## 2     35033 Chinook salmon mixed age juvenile length_mm   444
## 3     35034 Sockeye salmon juvenile           length_mm   139
## 4     35035 Sockeye salmon juvenile           length_mm   121
## 5     35036 Sockeye salmon juvenile           length_mm   112
## 6     35037 Steelhead      juvenile           length_mm   220
```


```r
head(df3b)
```

```
## # A tibble: 6 x 5
##   salmon_id common_name    age_classbylength  variable   value
##       &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;              &lt;fct&gt;      &lt;dbl&gt;
## 1     35032 Chinook salmon yearling           IGF1_ng_ml  41.3
## 2     35033 Chinook salmon mixed age juvenile IGF1_ng_ml  62.1
## 3     35034 Sockeye salmon juvenile           IGF1_ng_ml  NA  
## 4     35035 Sockeye salmon juvenile           IGF1_ng_ml  NA  
## 5     35036 Sockeye salmon juvenile           IGF1_ng_ml  NA  
## 6     35037 Steelhead      juvenile           IGF1_ng_ml  42.7
```

---
## What is a tidy dataset?

A tidy dataset is a data frame (or table) for which the following are true:

* Each variable has its own column
* Each observation has its own row
* Each value has its own cell

***Which of our dataframes is tidy?***
--
&lt;p&gt;&amp;nbsp;&lt;/p&gt;
Our first dataframe is tidy

---
## Why bother?

Consistent dataframe layouts help to ensure that all values are present and that relationships between data points are clear.

R is a vectorized programming language. R builds data frames from vectors, and R works best when its operation are vectorized. Tidy data utilizes of both of these aspects of R.

=&gt; Precise and Fast

---
## Lets load in the tidyverse

```r
library(tidyverse)
```

---
# dplyr
This package contains a variety of tools to access and manipulate dataframes.  

They have a common rationale:

![dplyr](./imgs/dplyr.png)

---
## dplyr::select - (pt1)
**Select** allows you to make a vector from a specific variable

```r
# Select one variable (common_name)
select(df1, common_name)
```

```
## # A tibble: 97 x 1
##    common_name   
##    &lt;chr&gt;         
##  1 Chinook salmon
##  2 Sockeye salmon
##  3 Sockeye salmon
##  4 Steelhead     
##  5 Steelhead     
##  6 Chinook salmon
##  7 Sockeye salmon
##  8 Steelhead     
##  9 Steelhead     
## 10 Steelhead     
## # … with 87 more rows
```

---
## dplyr::select - (pt2)
**Select** allows you to make a dataframe from several variables

```r
# Select two variables (age_classbylength and common_name)
select(df1, age_classbylength, common_name)
```

```
## # A tibble: 97 x 2
##    age_classbylength  common_name   
##    &lt;chr&gt;              &lt;chr&gt;         
##  1 yearling           Chinook salmon
##  2 juvenile           Sockeye salmon
##  3 juvenile           Sockeye salmon
##  4 juvenile           Steelhead     
##  5 juvenile           Steelhead     
##  6 mixed age juvenile Chinook salmon
##  7 juvenile           Sockeye salmon
##  8 juvenile           Steelhead     
##  9 juvenile           Steelhead     
## 10 juvenile           Steelhead     
## # … with 87 more rows
```

---
## dplyr::select - (pt3)
**Select** allows you to make a dataframe excluding a variable

```r
# Select all but one variable (length_mm)
select(df1,-length_mm)
```

```
## # A tibble: 97 x 4
##    salmon_id common_name    age_classbylength  IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                   &lt;dbl&gt;
##  1     35032 Chinook salmon yearling                 41.3
##  2     35035 Sockeye salmon juvenile                 NA  
##  3     35036 Sockeye salmon juvenile                 NA  
##  4     35037 Steelhead      juvenile                 42.7
##  5     35038 Steelhead      juvenile                 NA  
##  6     35033 Chinook salmon mixed age juvenile       62.1
##  7     35034 Sockeye salmon juvenile                 NA  
##  8     35048 Steelhead      juvenile                 24.2
##  9     35049 Steelhead      juvenile                 NA  
## 10     35050 Steelhead      juvenile                 63.5
## # … with 87 more rows
```

---
## dplyr::select - (pt4)
**Select** allows you to make a dataframe from a range of variables

```r
# Select all a range of contiguous varibles (common_name:length_mm)
select(df1, common_name:length_mm)
```

```
## # A tibble: 97 x 3
##    common_name    age_classbylength  length_mm
##    &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;
##  1 Chinook salmon yearling                 147
##  2 Sockeye salmon juvenile                 121
##  3 Sockeye salmon juvenile                 112
##  4 Steelhead      juvenile                 220
##  5 Steelhead      juvenile                 152
##  6 Chinook salmon mixed age juvenile       444
##  7 Sockeye salmon juvenile                 139
##  8 Steelhead      juvenile                 288
##  9 Steelhead      juvenile                 190
## 10 Steelhead      juvenile                 283
## # … with 87 more rows
```

---
## dplyr::filter - (pt1)
__Filter__ allows you to access observations based on a criteria

```r
# Filter all observation where the variable common_name is Sockeye salmon
filter(df1, common_name == 'Sockeye salmon')
```

```
## # A tibble: 11 x 5
##    salmon_id common_name    age_classbylength length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                 &lt;dbl&gt;      &lt;dbl&gt;
##  1     35035 Sockeye salmon juvenile                121         NA
##  2     35036 Sockeye salmon juvenile                112         NA
##  3     35034 Sockeye salmon juvenile                139         NA
##  4     35144 Sockeye salmon juvenile                140         NA
##  5     35147 Sockeye salmon juvenile                115         NA
##  6     35096 Sockeye salmon juvenile                115         NA
##  7     35097 Sockeye salmon juvenile                110         NA
##  8     35098 Sockeye salmon juvenile                112         NA
##  9     35099 Sockeye salmon juvenile                111         NA
## 10     35100 Sockeye salmon juvenile                118         NA
## 11     35119 Sockeye salmon juvenile                122         NA
```

---
## dplyr::filter - (pt2)
__Filter__ allows you to access observations based on several criteria

```r
# Filter all observations where the variable common_name is either Sockeye salmon or Chinook Salmon
filter(df1, common_name %in% c('Sockeye salmon', 'Chinook salmon'))
```

```
## # A tibble: 57 x 5
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
##  1     35032 Chinook salmon yearling                 147       41.3
##  2     35035 Sockeye salmon juvenile                 121       NA  
##  3     35036 Sockeye salmon juvenile                 112       NA  
##  4     35033 Chinook salmon mixed age juvenile       444       62.1
##  5     35034 Sockeye salmon juvenile                 139       NA  
##  6     35142 Chinook salmon yearling                 149       66.5
##  7     35143 Chinook salmon yearling                 204       80.9
##  8     35144 Sockeye salmon juvenile                 140       NA  
##  9     35145 Chinook salmon yearling                 130       23.4
## 10     35146 Chinook salmon mixed age juvenile       422      101. 
## # … with 47 more rows
```

---
## dplyr::filter - (pt3)
__Filter__ allows you to access observations based on specific strings

```r
# Filter all observations where the variable common_name ends with 'salmon'. To do this we use stringr function str_ends recognise strings that end with 'salmon'.
filter(df1, str_ends(common_name, 'salmon'))
```

```
## # A tibble: 59 x 5
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
##  1     35032 Chinook salmon yearling                 147       41.3
##  2     35035 Sockeye salmon juvenile                 121       NA  
##  3     35036 Sockeye salmon juvenile                 112       NA  
##  4     35033 Chinook salmon mixed age juvenile       444       62.1
##  5     35034 Sockeye salmon juvenile                 139       NA  
##  6     35142 Chinook salmon yearling                 149       66.5
##  7     35143 Chinook salmon yearling                 204       80.9
##  8     35144 Sockeye salmon juvenile                 140       NA  
##  9     35145 Chinook salmon yearling                 130       23.4
## 10     35146 Chinook salmon mixed age juvenile       422      101. 
## # … with 49 more rows
```

---
## dplyr::filter - (pt4)
__Filter__ allows you to access observations based on operators

```r
# Filter all observations where the variable length_mm is greater than 200 or less than 120
filter(df1, length_mm &gt; 200 | length_mm &lt; 120)
```

```
## # A tibble: 36 x 5
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
##  1     35036 Sockeye salmon juvenile                 112       NA  
##  2     35037 Steelhead      juvenile                 220       42.7
##  3     35033 Chinook salmon mixed age juvenile       444       62.1
##  4     35048 Steelhead      juvenile                 288       24.2
##  5     35050 Steelhead      juvenile                 283       63.5
##  6     35051 Steelhead      juvenile                 279       61.2
##  7     35052 Steelhead      juvenile                 235       30.6
##  8     35053 Steelhead      juvenile                 230       49.4
##  9     35056 Steelhead      juvenile                 208       57.4
## 10     35057 Steelhead      juvenile                 240       20.2
## # … with 26 more rows
```

---
## CHALLENGE: Time for an exercise!

Exercise on dplyr's select and filter can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/exercises/EX1_RU_tidyverse_dplyr_select_and_filter.html)

---
### Answers to exercise.

Answers can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/exercises/EX1_ANSWERS_RU_tidyverse_dplyr_select_and_filter.html)

R code for solutions can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/exercises/EX1_ANSWERS_RU_tidyverse_dplyr_select_and_filter.R)

---
## dplyr::arrange - (pt1)
_Arrange_ sorts the dataframe based on a specific variable

```r
# Arrange the data based on the variable length_mm
arrange(df1, length_mm)
```

```
## # A tibble: 97 x 5
##    salmon_id common_name    age_classbylength length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                 &lt;dbl&gt;      &lt;dbl&gt;
##  1     35095 Chinook salmon subyearling              90         NA
##  2     35097 Sockeye salmon juvenile                110         NA
##  3     35099 Sockeye salmon juvenile                111         NA
##  4     35036 Sockeye salmon juvenile                112         NA
##  5     35098 Sockeye salmon juvenile                112         NA
##  6     35147 Sockeye salmon juvenile                115         NA
##  7     35096 Sockeye salmon juvenile                115         NA
##  8     35100 Sockeye salmon juvenile                118         NA
##  9     35035 Sockeye salmon juvenile                121         NA
## 10     35119 Sockeye salmon juvenile                122         NA
## # … with 87 more rows
```
---
## dplyr::arrange - (pt2)
_Arrange_ sorts the dataframe based on specific variables

```r
# Arrange the data first based on the variable common_name, 
# then secondly based on length_mm in a descending order.
arrange(df1, common_name, desc(length_mm))
```

```
## # A tibble: 97 x 5
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
##  1     35033 Chinook salmon mixed age juvenile       444      62.1 
##  2     35146 Chinook salmon mixed age juvenile       422     101.  
##  3     35110 Chinook salmon mixed age juvenile       275      81.5 
##  4     35129 Chinook salmon yearling                 225      72.7 
##  5     35103 Chinook salmon yearling                 216      81.2 
##  6     35115 Chinook salmon yearling                 215      53.5 
##  7     35112 Chinook salmon yearling                 205      90.5 
##  8     35143 Chinook salmon yearling                 204      80.9 
##  9     35079 Chinook salmon yearling                 199      53.2 
## 10     35081 Chinook salmon yearling                 196       5.56
## # … with 87 more rows
```

---
## dplyr::mutate - (pt1)
_Mutate_ creates a new variable based on some form of computation

```r
# A new variable is created based on the calculation of the
# z-score of the variable IGF1_ng_ml using scale()
mutate(df1, scale(IGF1_ng_ml))
```

```
## # A tibble: 97 x 6
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml `scale(IGF1_ng_ml)`[,1]
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;                   &lt;dbl&gt;
##  1     35032 Chinook salmon yearling                 147       41.3                  -0.258
##  2     35035 Sockeye salmon juvenile                 121       NA                    NA    
##  3     35036 Sockeye salmon juvenile                 112       NA                    NA    
##  4     35037 Steelhead      juvenile                 220       42.7                  -0.191
##  5     35038 Steelhead      juvenile                 152       NA                    NA    
##  6     35033 Chinook salmon mixed age juvenile       444       62.1                   0.704
##  7     35034 Sockeye salmon juvenile                 139       NA                    NA    
##  8     35048 Steelhead      juvenile                 288       24.2                  -1.04 
##  9     35049 Steelhead      juvenile                 190       NA                    NA    
## 10     35050 Steelhead      juvenile                 283       63.5                   0.766
## # … with 87 more rows
```
---
## dplyr::mutate - (pt2)
_Mutate_ creates a named variable based on some form of computation

```r
# A new variable is created called IGFngml_zscore, based on the 
# calculation of the z-score of the variable IGF1_ng_ml using scale()
mutate(df1, IGFngml_zscore = scale(IGF1_ng_ml))
```

```
## # A tibble: 97 x 6
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml IGFngml_zscore[,1]
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;              &lt;dbl&gt;
##  1     35032 Chinook salmon yearling                 147       41.3             -0.258
##  2     35035 Sockeye salmon juvenile                 121       NA               NA    
##  3     35036 Sockeye salmon juvenile                 112       NA               NA    
##  4     35037 Steelhead      juvenile                 220       42.7             -0.191
##  5     35038 Steelhead      juvenile                 152       NA               NA    
##  6     35033 Chinook salmon mixed age juvenile       444       62.1              0.704
##  7     35034 Sockeye salmon juvenile                 139       NA               NA    
##  8     35048 Steelhead      juvenile                 288       24.2             -1.04 
##  9     35049 Steelhead      juvenile                 190       NA               NA    
## 10     35050 Steelhead      juvenile                 283       63.5              0.766
## # … with 87 more rows
```

---
## CHALLENGE: Time for an exercise!

Exercise on dplyr's arrange and mutate can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/exercises/EX1_RU_tidyverse_dplyr_arrange_and_mutate.html)

---
### Answers to exercise.

Answers can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/exercises/EX1_ANSWERS_RU_tidyverse_dplyr_arrange_and_mutate.html)

R code for solutions can be found [here](http://rockefelleruniversity.github.io/RU_tidyverse/tidyR/exercises/EX1_ANSWERS_RU_tidyverse_dplyr_arrange_and_mutate.R)

---
## dplyr::summarize - (pt1)
_Summarize_ applies aggregating or summary function to a group i.e. counting

```r
# First we define the common_name as a group. 
df1_byname &lt;- group_by(df1, common_name)

# Summarise is used to count over the grouped common_names
summarise(df1_byname, count = n())
```

```
## # A tibble: 4 x 2
##   common_name    count
##   &lt;chr&gt;          &lt;int&gt;
## 1 Chinook salmon    46
## 2 Coho salmon        2
## 3 Sockeye salmon    11
## 4 Steelhead         38
```
---
## dplyr::summarize - (pt2)
_Summarize_ applies aggregating or summary function to a group i.e. means

```r
# Summarise is used to calculate mean IGF1_ng_ml over the
# grouped common_names
summarise(df1_byname, IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm = T))
```

```
## # A tibble: 4 x 2
##   common_name    IGF1_ng_ml_ave
##   &lt;chr&gt;                   &lt;dbl&gt;
## 1 Chinook salmon           46.8
## 2 Coho salmon              73.6
## 3 Sockeye salmon          NaN  
## 4 Steelhead                46.1
```

---
## dplyr::group_by - (pt1)
_Grouping_ can also help filter within groups

```r
# Filter observations with the 2 smallest length_mm for each grouped common_names
filter(df1_byname, rank(length_mm) &lt;= 2)
```

```
## # A tibble: 8 x 5
## # Groups:   common_name [4]
##   salmon_id common_name    age_classbylength length_mm IGF1_ng_ml
##       &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                 &lt;dbl&gt;      &lt;dbl&gt;
## 1     35038 Steelhead      juvenile                152       NA  
## 2     35055 Steelhead      juvenile                123       55.7
## 3     35145 Chinook salmon yearling                130       23.4
## 4     35085 Coho salmon    yearling                140       NA  
## 5     35087 Coho salmon    yearling                164       73.6
## 6     35095 Chinook salmon subyearling              90       NA  
## 7     35097 Sockeye salmon juvenile                110       NA  
## 8     35099 Sockeye salmon juvenile                111       NA
```
---
## dplyr::group_by - (pt2)
_Grouping_ can also help filter within groups

```r
# Filter observations with at least 5 for each grouped common_names
filter(df1_byname, n() &gt; 5)
```

```
## # A tibble: 95 x 5
## # Groups:   common_name [3]
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
##  1     35032 Chinook salmon yearling                 147       41.3
##  2     35035 Sockeye salmon juvenile                 121       NA  
##  3     35036 Sockeye salmon juvenile                 112       NA  
##  4     35037 Steelhead      juvenile                 220       42.7
##  5     35038 Steelhead      juvenile                 152       NA  
##  6     35033 Chinook salmon mixed age juvenile       444       62.1
##  7     35034 Sockeye salmon juvenile                 139       NA  
##  8     35048 Steelhead      juvenile                 288       24.2
##  9     35049 Steelhead      juvenile                 190       NA  
## 10     35050 Steelhead      juvenile                 283       63.5
## # … with 85 more rows
```
---
## dplyr::group_by - (pt3)
_Grouping_ creates a new variable based on some form of computation within the group

```r
# A new variable is created using z-score within the grouped common_names
mutate(df1_byname, IGFngml_zscore = scale(IGF1_ng_ml))
```

```
## # A tibble: 97 x 6
## # Groups:   common_name [4]
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml IGFngml_zscore
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;          &lt;dbl&gt;
##  1     35032 Chinook salmon yearling                 147       41.3         -0.236
##  2     35035 Sockeye salmon juvenile                 121       NA           NA    
##  3     35036 Sockeye salmon juvenile                 112       NA           NA    
##  4     35037 Steelhead      juvenile                 220       42.7         -0.176
##  5     35038 Steelhead      juvenile                 152       NA           NA    
##  6     35033 Chinook salmon mixed age juvenile       444       62.1          0.652
##  7     35034 Sockeye salmon juvenile                 139       NA           NA    
##  8     35048 Steelhead      juvenile                 288       24.2         -1.12 
##  9     35049 Steelhead      juvenile                 190       NA           NA    
## 10     35050 Steelhead      juvenile                 283       63.5          0.890
## # … with 87 more rows
```

---
## CHALLENGE

1. Group df1 by the variable _age_class_. 
2. Filter to get the biggest 5 by the variable _length_ in each group. 
3. Summarise this data frame over the variable _length_ by calculating the mean. 

---
## ANSWERS


```r
# Group df1 by the variable _age_class_. 
df1_A &lt;- group_by(df1, age_classbylength)

# Filter to get the biggest 5 by the variable _length_ in each group. 
df1_A &lt;- filter(df1_A, rank(-length_mm) &lt;= 5)

# Summarise this data frame over the variable _length_ by calculating the mean.
summarise(df1_A, mean_length_mm = mean(length_mm, na.rm = T))
```

```
## # A tibble: 4 x 2
##   age_classbylength  mean_length_mm
##   &lt;chr&gt;                       &lt;dbl&gt;
## 1 juvenile                     274 
## 2 mixed age juvenile           380.
## 3 subyearling                   90 
## 4 yearling                     213
```

---
# Piping to string functions together

Piping was allows you to pass the result from one expression directly into another.  
magrittR package developed the %&gt;% pipe which is integral to the tidy way of formatting code

The pattern is similar but now follows a specific logical flow:

![dplyr](./imgs/dplyr_pipe.png)
---
## Piping vs. not piping

```r
# Without pipe
df1_byname &lt;- group_by(df1, common_name)
summarise(df1_byname, IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm=T))
```

```
## # A tibble: 4 x 2
##   common_name    IGF1_ng_ml_ave
##   &lt;chr&gt;                   &lt;dbl&gt;
## 1 Chinook salmon           46.8
## 2 Coho salmon              73.6
## 3 Sockeye salmon          NaN  
## 4 Steelhead                46.1
```

```r
# With pipe
df1 %&gt;% 
  group_by(common_name) %&gt;% 
  summarise(IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm=T))
```

```
## # A tibble: 4 x 2
##   common_name    IGF1_ng_ml_ave
##   &lt;chr&gt;                   &lt;dbl&gt;
## 1 Chinook salmon           46.8
## 2 Coho salmon              73.6
## 3 Sockeye salmon          NaN  
## 4 Steelhead                46.1
```

---
## Linking functions with pipes - (pt1)

```r
# (1) Group by common_name
# (2) Filter to all those that have length bigger then 200
# (3) Summarise is used to calculate mean IGF1_ng_ml over the grouped 
# common_names for these larger fish

df1 %&gt;%
  group_by(common_name) %&gt;% 
  filter(length_mm &gt; 200) %&gt;% 
  summarise(IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm = T))
```

```
## # A tibble: 2 x 2
##   common_name    IGF1_ng_ml_ave
##   &lt;chr&gt;                   &lt;dbl&gt;
## 1 Chinook salmon           77.9
## 2 Steelhead                45.3
```
---
## Linking functions with pipes - (pt2)

```r
# (1) Create new variable that is discrete label depending on size of the fish
# (2) Group by common_name and size
# (3) Summarise is used to calculate mean IGF1_ng_ml over the grouped 
# common_names and sizes

df1 %&gt;%
  mutate(size = if_else(length_mm &gt; 200, 'big_fish', 'small_fish')) %&gt;% 
  group_by(common_name, size) %&gt;% 
  summarise(IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm = T))
```

```
## # A tibble: 6 x 3
## # Groups:   common_name [4]
##   common_name    size       IGF1_ng_ml_ave
##   &lt;chr&gt;          &lt;chr&gt;               &lt;dbl&gt;
## 1 Chinook salmon big_fish             77.9
## 2 Chinook salmon small_fish           39.5
## 3 Coho salmon    small_fish           73.6
## 4 Sockeye salmon small_fish          NaN  
## 5 Steelhead      big_fish             45.3
## 6 Steelhead      small_fish           47.3
```
---
## Linking functions with pipes - (pt3)

```r
# (1) Create new variable that is discrete label depending on size of the fish
# (2) Group by common_name and size
# (3) Summarise is used to calculate mean IGF1_ng_ml over the grouped 
# common_names and sizes
# (4) Filter out Coho and Sockeye salmon

df1 %&gt;%
  mutate(size = if_else(length_mm &gt; 200, 'big_fish', 'small_fish')) %&gt;% 
  group_by(common_name, size) %&gt;% 
  summarise(IGF1_ng_ml_ave = mean(IGF1_ng_ml, na.rm = T)) %&gt;%
  filter(common_name != 'Coho salmon')  %&gt;%
  filter(common_name !='Sockeye salmon')
```

```
## # A tibble: 4 x 3
## # Groups:   common_name [2]
##   common_name    size       IGF1_ng_ml_ave
##   &lt;chr&gt;          &lt;chr&gt;               &lt;dbl&gt;
## 1 Chinook salmon big_fish             77.9
## 2 Chinook salmon small_fish           39.5
## 3 Steelhead      big_fish             45.3
## 4 Steelhead      small_fish           47.3
```

---
## You can pipe straight to plots - (pt1)

```r
p &lt;- mutate(df1,size=if_else(length_mm&gt;200, 'big_fish', 'small_fish')) %&gt;% 
  group_by(common_name, size) %&gt;% 
  summarize(IGF1_ng_ml_ave=mean(IGF1_ng_ml, na.rm=T)) %&gt;% 
  filter(common_name != 'Coho salmon')  %&gt;% 
  filter(common_name != 'Sockeye salmon') %&gt;% 
  ggplot(aes(x = common_name, y = IGF1_ng_ml_ave, group = size, fill = size)) +
  geom_bar(stat = "identity", position = "dodge") + 
  theme(axis.text.x = element_text(angle = 90)) +
  scale_fill_brewer(palette = "Paired")
```
---
## You can pipe straight to plots - (pt2)

```r
p
```

![](20191009_tidyr_files/figure-html/unnamed-chunk-30-1.png)&lt;!-- --&gt;

---
## CHALLENGE

1. Filter to only look at _juvenile_ in variable 
2. Group by common_name
3. Create new variable that is z-score of _length_ across groups
4. Create boxplot of grouped length zscores

---
## ANSWER

```r
df1 %&gt;% 
  filter(age_classbylength == 'yearling') %&gt;%
  group_by(common_name) %&gt;% 
  mutate(length_zscore = scale(length_mm)) %&gt;%
  ggplot(aes(x = common_name, y = length_zscore)) +
  geom_boxplot(fill=c('gold','darkorange'))
```

![](20191009_tidyr_files/figure-html/unnamed-chunk-31-1.png)&lt;!-- --&gt;

---
# readr: Reading data into R
So we blasted through what being tidy can give you. 
Now lets start from the beginning and tidy some data. First step is to read in data. 

readr:

* read_csv(): comma separated (CSV) files  
* read_tsv(): tab separated files  
* read_delim(): general delimited files  
* read_fwf(): fixed width files  
* read_table(): tabular files where columns are separated by white-space  
* read_log(): web log files   

---
## Reading in with base

```r
untidy_counts_base &lt;- read.csv("dataset/hemato_rnaseq_counts.csv")
# Base will print out everything
untidy_counts_base
```

```
##     ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
## 1      350    204       0    103       0
## 2      351  15586     479  10476      39
## 3      353    842     355   1188      86
## 4      354      0       0      0       0
## 5      355    123     291    139      16
## 6      356      1       1      0       0
## 7      357    380       3    177       0
## 8      358    572    2225    597    4051
## 9      359      0      12      1       0
## 10     360    320     502     46    1114
## 11     361      0       1      0       0
## 12     362      3       1     15       0
## 13     363     14       6      4       1
## 14     364      7       0      1       0
## 15     366      6       0      1       0
## 16     367     42       0     51       1
## 17     368     28       0     24       0
## 18     369   1204    1034    833     478
## 19     372   2829    1864   2741     771
## 20     373    179     728    148     795
## 21     374     76       5    138       2
## 22     375   4428    6697   4970    4328
## 23     377   3170     314   2576      11
## 24     378   1839    4845   1767    2975
## 25     379    178       1    181       0
## 26     381   1617     574   1159     339
## 27     382   2874    2265   1746    1668
## 28     383     63    1632     40     721
## 29     384    148   10977    118      94
## 30     387   8899    2457   7405    1228
## 31     388  12598     171   5090      70
## 32     389   2709     193   2313       5
## 33     390   1004       0    395       0
## 34     391   1038     577   1176     164
## 35     392   1527     304    786      71
## 36     393   2949     138   1540       3
## 37     394   1525     464   1062     134
## 38     395    348      67    123       0
## 39     396   6503     702   4723     169
## 40     397  12997     410  11265      38
## 41     398      0       0      0       0
## 42     399    223      11    422       2
## 43     400   1188     147    806      56
## 44     401      0       0      0       0
## 45     402    504     218    496      80
## 46     403    289      25    166       4
## 47     405   1481     824   1004     812
## 48     406    295     175     87      35
## 49     407      4       1      2       1
## 50     408   2451     111   1523       6
## 51     409   2480    1819   1356     226
## 52     410    433     197    215      77
## 53     411    829     217    441     131
## 54     412    312      45    138      17
## 55     414    516      15    396       9
## 56     415     20       0     13       0
## 57     416      2       1      4       0
## 58     417      0       0      0       0
## 59     419      6       5      2       7
## 60     420    141    1136     94    1217
## 61     421    213     255     93     208
## 62     427   4699   11889   1729     926
## 63     429      0       0      2       0
## 64     430     34      13     22       0
## 65     432     63       0     55       1
## 66     433     38       0     26       0
## 67     434      1       1      2       0
## 68     435    408     151    284      34
## 69     440    157    2520    111     535
## 70     443      5       0      4       0
## 71     444   1583     151    747      14
## 72     445    116      15     90       1
## 73     460     34       0     68       0
## 74     462     70       0     31       1
## 75     463   1244     118    492      71
## 76     466    538     480    393     218
## 77     467   2506     402   2130      18
## 78     468   7991    6132   5307    1883
## 79     471   1272     771   1392      53
## 80     472   1389     628    739     138
## 81     473   4173     783   1901     776
## 82     474      0       0      0       0
## 83     475    284      83    467      34
## 84     476   4952    3453   4202    1416
## 85     477     13       3     26       2
## 86     478     78     121     67       0
## 87     479     17       0      4       0
## 88     480      9       5     11       1
## 89     481   1937      18   1017      34
## 90     482    157    1392     75    1660
## 91     483   1075    1454   1789    1141
## 92     486     47       0     18       0
## 93     487     29      33     19       3
## 94     488   4529    1118   2925     269
## 95     489   3465     153   3188       8
## 96     490   1610    1263    913     665
## 97     491     12       1      4       0
## 98     492      4       0      6       0
## 99     493   5011    3585   3053     743
## 100    495      0       0      0       0
```
---
## Reading in with readr - (pt1)

```r
# readr gives you a tibble. 
read_csv("dataset/hemato_rnaseq_counts.csv")
```

```
## Parsed with column specification:
## cols(
##   ENTREZ = col_double(),
##   CD34_1 = col_double(),
##   ORTHO_1 = col_double(),
##   CD34_2 = col_double(),
##   ORTHO_2 = col_double()
## )
```

```
## # A tibble: 100 x 5
##    ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
##     &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;
##  1    350    204       0    103       0
##  2    351  15586     479  10476      39
##  3    353    842     355   1188      86
##  4    354      0       0      0       0
##  5    355    123     291    139      16
##  6    356      1       1      0       0
##  7    357    380       3    177       0
##  8    358    572    2225    597    4051
##  9    359      0      12      1       0
## 10    360    320     502     46    1114
## # … with 90 more rows
```
---
## Reading in with readr - (pt2)

```r
# Tibbles carry and display extra information. 
# While reading in it is easy to specify data type. 
untidy_counts &lt;- read_csv("dataset/hemato_rnaseq_counts.csv", col_types = cols(
    ENTREZ = col_character(),
    CD34_1 = col_integer(),
    ORTHO_1 = col_integer(),
    CD34_2 = col_integer(),
    ORTHO_2 = col_integer()
  ))
untidy_counts
```

```
## # A tibble: 100 x 5
##    ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
##    &lt;chr&gt;   &lt;int&gt;   &lt;int&gt;  &lt;int&gt;   &lt;int&gt;
##  1 350       204       0    103       0
##  2 351     15586     479  10476      39
##  3 353       842     355   1188      86
##  4 354         0       0      0       0
##  5 355       123     291    139      16
##  6 356         1       1      0       0
##  7 357       380       3    177       0
##  8 358       572    2225    597    4051
##  9 359         0      12      1       0
## 10 360       320     502     46    1114
## # … with 90 more rows
```

---
## Subsetting to make tibbles - (pt1)

```r
# Can use the same way as base to interact with the 
# tibble dataframe - columns
untidy_counts[,1]
```

```
## # A tibble: 100 x 1
##    ENTREZ
##    &lt;chr&gt; 
##  1 350   
##  2 351   
##  3 353   
##  4 354   
##  5 355   
##  6 356   
##  7 357   
##  8 358   
##  9 359   
## 10 360   
## # … with 90 more rows
```
---
## Subsetting to make tibbles - (pt2)

```r
# Can use the same way as base to interact with the
# tibble dataframe - rows
untidy_counts[1,]
```

```
## # A tibble: 1 x 5
##   ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
##   &lt;chr&gt;   &lt;int&gt;   &lt;int&gt;  &lt;int&gt;   &lt;int&gt;
## 1 350       204       0    103       0
```
---
## Subsetting to make tibbles - (pt3)

```r
# Can also not specify which dimension you pull from. 
# This will default to grabbing the column
untidy_counts[1]
```

```
## # A tibble: 100 x 1
##    ENTREZ
##    &lt;chr&gt; 
##  1 350   
##  2 351   
##  3 353   
##  4 354   
##  5 355   
##  6 356   
##  7 357   
##  8 358   
##  9 359   
## 10 360   
## # … with 90 more rows
```
---
## Subsetting to make vectors - (pt1)

```r
# All the prior outputs have been outputting another tibble. 
# If double brackets are used a vector is returned 
untidy_counts[[1]]
```

```
##   [1] "350" "351" "353" "354" "355" "356" "357" "358" "359" "360" "361" "362" "363" "364" "366" "367" "368" "369" "372" "373" "374" "375"
##  [23] "377" "378" "379" "381" "382" "383" "384" "387" "388" "389" "390" "391" "392" "393" "394" "395" "396" "397" "398" "399" "400" "401"
##  [45] "402" "403" "405" "406" "407" "408" "409" "410" "411" "412" "414" "415" "416" "417" "419" "420" "421" "427" "429" "430" "432" "433"
##  [67] "434" "435" "440" "443" "444" "445" "460" "462" "463" "466" "467" "468" "471" "472" "473" "474" "475" "476" "477" "478" "479" "480"
##  [89] "481" "482" "483" "486" "487" "488" "489" "490" "491" "492" "493" "495"
```
---
## Subsetting to make vectors - (pt2)

```r
# This is also the case if you use the dollar and colname 
# to access a column
untidy_counts$ENTREZ
```

```
##   [1] "350" "351" "353" "354" "355" "356" "357" "358" "359" "360" "361" "362" "363" "364" "366" "367" "368" "369" "372" "373" "374" "375"
##  [23] "377" "378" "379" "381" "382" "383" "384" "387" "388" "389" "390" "391" "392" "393" "394" "395" "396" "397" "398" "399" "400" "401"
##  [45] "402" "403" "405" "406" "407" "408" "409" "410" "411" "412" "414" "415" "416" "417" "419" "420" "421" "427" "429" "430" "432" "433"
##  [67] "434" "435" "440" "443" "444" "445" "460" "462" "463" "466" "467" "468" "471" "472" "473" "474" "475" "476" "477" "478" "479" "480"
##  [89] "481" "482" "483" "486" "487" "488" "489" "490" "491" "492" "493" "495"
```

---
## CHALLENGE

* Read in count data as different data types. What happens when we extract the vector of these counts?

---
## ANSWER

```r
untidy_counts_test &lt;- read_csv("dataset/hemato_rnaseq_counts.csv", col_types = cols(
    ENTREZ = col_character(),
    CD34_1 = col_integer(),
    ORTHO_1 = col_character(),
    CD34_2 = col_factor(),
    ORTHO_2 = col_logical()
  ))
```

```
## Warning: 62 parsing failures.
## row     col           expected actual                               file
##   2 ORTHO_2 1/0/T/F/TRUE/FALSE   39   'dataset/hemato_rnaseq_counts.csv'
##   3 ORTHO_2 1/0/T/F/TRUE/FALSE   86   'dataset/hemato_rnaseq_counts.csv'
##   5 ORTHO_2 1/0/T/F/TRUE/FALSE   16   'dataset/hemato_rnaseq_counts.csv'
##   8 ORTHO_2 1/0/T/F/TRUE/FALSE   4051 'dataset/hemato_rnaseq_counts.csv'
##  10 ORTHO_2 1/0/T/F/TRUE/FALSE   1114 'dataset/hemato_rnaseq_counts.csv'
## ... ....... .................. ...... ..................................
## See problems(...) for more details.
```

```r
untidy_counts_test$CD34_1 %&gt;% head(n=8)
```

```
## [1]   204 15586   842     0   123     1   380   572
```

```r
untidy_counts_test$ORTHO_1 %&gt;% head(n=8)
```

```
## [1] "0"    "479"  "355"  "0"    "291"  "1"    "3"    "2225"
```

```r
untidy_counts_test$CD34_2 %&gt;% head(n=8)
```

```
## [1] 103   10476 1188  0     139   0     177   597  
## 82 Levels: 103 10476 1188 0 139 177 597 1 46 15 4 51 24 833 2741 148 138 4970 2576 1767 181 1159 1746 40 118 7405 5090 2313 395 ... 3053
```

```r
untidy_counts_test$ORTHO_2 %&gt;% head(n=8)
```

```
## [1] FALSE    NA    NA FALSE    NA FALSE FALSE    NA
```

---
## Tibbles: Converting to tibble - (pt1)

```r
# Can convert base dataframes into tibbles
as_tibble(untidy_counts_base)
```

```
## # A tibble: 100 x 5
##    ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
##     &lt;int&gt;  &lt;int&gt;   &lt;int&gt;  &lt;int&gt;   &lt;int&gt;
##  1    350    204       0    103       0
##  2    351  15586     479  10476      39
##  3    353    842     355   1188      86
##  4    354      0       0      0       0
##  5    355    123     291    139      16
##  6    356      1       1      0       0
##  7    357    380       3    177       0
##  8    358    572    2225    597    4051
##  9    359      0      12      1       0
## 10    360    320     502     46    1114
## # … with 90 more rows
```
---
## Tibbles: Converting to tibble - (pt2)

```r
# Once it is a tibble it is straight forward to modify the datatype
untidy_counts_base &lt;- as_tibble(untidy_counts_base) %&gt;%
  mutate_at(vars(ENTREZ), as.character)
untidy_counts_base
```

```
## # A tibble: 100 x 5
##    ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
##    &lt;chr&gt;   &lt;int&gt;   &lt;int&gt;  &lt;int&gt;   &lt;int&gt;
##  1 350       204       0    103       0
##  2 351     15586     479  10476      39
##  3 353       842     355   1188      86
##  4 354         0       0      0       0
##  5 355       123     291    139      16
##  6 356         1       1      0       0
##  7 357       380       3    177       0
##  8 358       572    2225    597    4051
##  9 359         0      12      1       0
## 10 360       320     502     46    1114
## # … with 90 more rows
```
---
## Tibbles: Converting from tibble

```r
# Some tools are not tibble friendly. Calling as.data.frame is 
# sufficient to convert it back to a base data frame
as.data.frame(untidy_counts_base) %&gt;% head(n=12)
```

```
##    ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
## 1     350    204       0    103       0
## 2     351  15586     479  10476      39
## 3     353    842     355   1188      86
## 4     354      0       0      0       0
## 5     355    123     291    139      16
## 6     356      1       1      0       0
## 7     357    380       3    177       0
## 8     358    572    2225    597    4051
## 9     359      0      12      1       0
## 10    360    320     502     46    1114
## 11    361      0       1      0       0
## 12    362      3       1     15       0
```

---
## Tibbles: Make your own - (pt1)
__We will make our own tibble now from scratch, using some metadata that will be useful later__

```r
# Lets load in some packages
library(org.Hs.eg.db)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)

# Lets use the ENTREZ ID as a key
keys &lt;- untidy_counts$ENTREZ

# We can use the ENTREZ ID to look up Gene Symbol
symbols &lt;- select(org.Hs.eg.db, keys = keys,columns = "SYMBOL", keytype = "ENTREZID")

# We can use the ENTREZ ID to look up the chormosome the gene resides on
chrs &lt;- select(TxDb.Hsapiens.UCSC.hg19.knownGene, keys = keys, columns = "TXCHROM", keytype = "GENEID")

# We can use the ENTREZ ID to get a list of genes with grange of their exons
geneExons &lt;- exonsBy(TxDb.Hsapiens.UCSC.hg19.knownGene,by = "gene")[keys]
```


---
## Tibbles: Make your own - (pt2)

```r
# We will then use an apply to get the transcript length from each gene in the 
# list. The transcript length is calculated by first flattening overlapping 
# exons with reduce(), then calculating the length of each exon with width(), 
# then summing upthe total exon length to get our transcript length. 
txsLength &lt;- sapply(geneExons, function(x){ x %&gt;%
    GenomicRanges::reduce() %&gt;%
    width() %&gt;%
    sum() })

# Finally we have all this metadata. Lets put it together into a tibble. 
counts_metadata &lt;- tibble(ID = symbols$ENTREZID, SYMBOL = symbols$SYMBOL, CHR = chrs$TXCHROM, LENGTH = txsLength)
```

# Finally we have all this metadata. Lets put it together into a tibble. 
counts_metadata &lt;- tibble(ID = symbols$ENTREZID, SYMBOL = symbols$SYMBOL, CHR = chrs$TXCHROM, LENGTH = txsLength)
---
## Tibbles: Make your own - (pt3)

```r
counts_metadata
```

```
## # A tibble: 100 x 4
##    ID    SYMBOL  CHR   LENGTH
##    &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;int&gt;
##  1 350   APOH    chr17   1201
##  2 351   APP     chr21   4480
##  3 353   APRT    chr16    807
##  4 354   KLK3    chr19   1906
##  5 355   FAS     chr10   6691
##  6 356   FASLG   chr1    1859
##  7 357   SHROOM2 chrX    8206
##  8 358   AQP1    chr7    3786
##  9 359   AQP2    chr12   4179
## 10 360   AQP3    chr9    2950
## # … with 90 more rows
```

---
# Tidying data up

What is wrong with the count dataframe from a tidy viewpoint?  

_Remember_. These are the rules:

* Each variable has its own column
* Each observation has its own row
* Each value has its own cell


```r
untidy_counts
```

```
## # A tibble: 100 x 5
##    ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
##    &lt;chr&gt;   &lt;int&gt;   &lt;int&gt;  &lt;int&gt;   &lt;int&gt;
##  1 350       204       0    103       0
##  2 351     15586     479  10476      39
##  3 353       842     355   1188      86
##  4 354         0       0      0       0
##  5 355       123     291    139      16
##  6 356         1       1      0       0
##  7 357       380       3    177       0
##  8 358       572    2225    597    4051
##  9 359         0      12      1       0
## 10 360       320     502     46    1114
## # … with 90 more rows
```

---
## Tidying data up

What is wrong with the count dataframe from a tidy viewpoint?  

_Remember_, these are the rules:

* Each variable has its own column
* Each observation has its own row
* Each value has its own cell

__A single variable with multiple columns__


```r
untidy_counts
```

```
## # A tibble: 100 x 5
##    ENTREZ CD34_1 ORTHO_1 CD34_2 ORTHO_2
##    &lt;chr&gt;   &lt;int&gt;   &lt;int&gt;  &lt;int&gt;   &lt;int&gt;
##  1 350       204       0    103       0
##  2 351     15586     479  10476      39
##  3 353       842     355   1188      86
##  4 354         0       0      0       0
##  5 355       123     291    139      16
##  6 356         1       1      0       0
##  7 357       380       3    177       0
##  8 358       572    2225    597    4051
##  9 359         0      12      1       0
## 10 360       320     502     46    1114
## # … with 90 more rows
```

---
## Reshape data with gather and spread

When all the content is in the data frame, but it is in the wrong orientation, tidyr has some tools to move the data around quickly and easily:  

__gather and spread__

---
## tidyr::gather 
Gather allows you to collapse single varibles that are spread over multiple columns

```r
# New columns Sample and Counts are created from whole dataframe
tidier_counts &lt;- gather(untidy_counts, key="Sample", value="counts", -ENTREZ)
tidier_counts
```

```
## # A tibble: 400 x 3
##    ENTREZ Sample counts
##    &lt;chr&gt;  &lt;chr&gt;   &lt;int&gt;
##  1 350    CD34_1    204
##  2 351    CD34_1  15586
##  3 353    CD34_1    842
##  4 354    CD34_1      0
##  5 355    CD34_1    123
##  6 356    CD34_1      1
##  7 357    CD34_1    380
##  8 358    CD34_1    572
##  9 359    CD34_1      0
## 10 360    CD34_1    320
## # … with 390 more rows
```
---
## tidyr::spread
Spread allows you to spread single variables over multiple columns

```r
spread(tidier_counts, key="Sample", value="counts")
```

```
## # A tibble: 100 x 5
##    ENTREZ CD34_1 CD34_2 ORTHO_1 ORTHO_2
##    &lt;chr&gt;   &lt;int&gt;  &lt;int&gt;   &lt;int&gt;   &lt;int&gt;
##  1 350       204    103       0       0
##  2 351     15586  10476     479      39
##  3 353       842   1188     355      86
##  4 354         0      0       0       0
##  5 355       123    139     291      16
##  6 356         1      0       1       0
##  7 357       380    177       3       0
##  8 358       572    597    2225    4051
##  9 359         0      1      12       0
## 10 360       320     46     502    1114
## # … with 90 more rows
```
---
## tidyr::pivot_longer/pivot_wider
Gather and spread are being retired. They are being replaced with pivot_tools.  
There is slightly different nomencalture, but they pretty much do the same thing.

```r
#
pivot_longer(untidy_counts,names_to = "Sample", values_to = "counts", cols = c(-ENTREZ))
gather(untidy_counts, key=Sample, value=counts, -ENTREZ)
# 
pivot_wider(tidier_counts, names_from = c(Sample), values_from = counts)
spread(tidier_counts, key=Sample, value=counts)
```

---
## CHALLENGE

* tidy up df2

---
## ANSWER

```r
df2 %&gt;% 
  spread(value = value, key = variable)
```

```
## # A tibble: 97 x 5
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
##  1     35032 Chinook salmon yearling                 147      41.3 
##  2     35033 Chinook salmon mixed age juvenile       444      62.1 
##  3     35034 Sockeye salmon juvenile                 139      NA   
##  4     35035 Sockeye salmon juvenile                 121      NA   
##  5     35036 Sockeye salmon juvenile                 112      NA   
##  6     35037 Steelhead      juvenile                 220      42.7 
##  7     35038 Steelhead      juvenile                 152      NA   
##  8     35039 Steelhead      juvenile                 201       9.92
##  9     35040 Steelhead      juvenile                 200      38.1 
## 10     35042 Steelhead      juvenile                 247      46.4 
## # … with 87 more rows
```

---
## What is next to tidy?

_Remember_, these are the rules:

* Each variable has its own column
* Each observation has its own row
* Each value has its own cell


```r
tidier_counts
```

```
## # A tibble: 400 x 3
##    ENTREZ Sample counts
##    &lt;chr&gt;  &lt;chr&gt;   &lt;int&gt;
##  1 350    CD34_1    204
##  2 351    CD34_1  15586
##  3 353    CD34_1    842
##  4 354    CD34_1      0
##  5 355    CD34_1    123
##  6 356    CD34_1      1
##  7 357    CD34_1    380
##  8 358    CD34_1    572
##  9 359    CD34_1      0
## 10 360    CD34_1    320
## # … with 390 more rows
```

---
## What is next to tidy?

_Remember_, these are the rules:

* Each variable has its own column
* Each observation has its own row
* Each value has its own cell

__Multiple variables in a single column__


```r
tidier_counts
```

```
## # A tibble: 400 x 3
##    ENTREZ Sample counts
##    &lt;chr&gt;  &lt;chr&gt;   &lt;int&gt;
##  1 350    CD34_1    204
##  2 351    CD34_1  15586
##  3 353    CD34_1    842
##  4 354    CD34_1      0
##  5 355    CD34_1    123
##  6 356    CD34_1      1
##  7 357    CD34_1    380
##  8 358    CD34_1    572
##  9 359    CD34_1      0
## 10 360    CD34_1    320
## # … with 390 more rows
```

---
## Splitting and combining varaibles

When there are several variables crammed into a single column, tidyr can be used to split single values into several:  

__separate__

---
## tidyr::separate - (pt1)

```r
# Separate allows you to break a strings in a variable by a separator.
# In this case the cell type and replicate number are broken by underscore
tidier_counts &lt;- separate(tidier_counts, Sample, sep = "_", into=c("CellType","Rep"), remove=TRUE)
tidier_counts
```

```
## # A tibble: 400 x 4
##    ENTREZ CellType Rep   counts
##    &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;
##  1 350    CD34     1        204
##  2 351    CD34     1      15586
##  3 353    CD34     1        842
##  4 354    CD34     1          0
##  5 355    CD34     1        123
##  6 356    CD34     1          1
##  7 357    CD34     1        380
##  8 358    CD34     1        572
##  9 359    CD34     1          0
## 10 360    CD34     1        320
## # … with 390 more rows
```
---
## tidyr::unite

```r
# Unite can go the other way. This can sometime be useful i.e. if you want a specific sample ID
unite(tidier_counts, Sample, CellType, Rep, remove=FALSE)
```

```
## # A tibble: 400 x 5
##    ENTREZ Sample CellType Rep   counts
##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;
##  1 350    CD34_1 CD34     1        204
##  2 351    CD34_1 CD34     1      15586
##  3 353    CD34_1 CD34     1        842
##  4 354    CD34_1 CD34     1          0
##  5 355    CD34_1 CD34     1        123
##  6 356    CD34_1 CD34     1          1
##  7 357    CD34_1 CD34     1        380
##  8 358    CD34_1 CD34     1        572
##  9 359    CD34_1 CD34     1          0
## 10 360    CD34_1 CD34     1        320
## # … with 390 more rows
```
---
## tidyr::separate - (pt2)

```r
# Remember you can always pipe everything together into a single expression
tidy_counts &lt;- untidy_counts %&gt;% 
  gather(key=Sample, value=counts, -ENTREZ) %&gt;% 
  separate(Sample, sep = "_", into = c("CellType","Rep"), remove=FALSE)
tidy_counts
```

```
## # A tibble: 400 x 5
##    ENTREZ Sample CellType Rep   counts
##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;
##  1 350    CD34_1 CD34     1        204
##  2 351    CD34_1 CD34     1      15586
##  3 353    CD34_1 CD34     1        842
##  4 354    CD34_1 CD34     1          0
##  5 355    CD34_1 CD34     1        123
##  6 356    CD34_1 CD34     1          1
##  7 357    CD34_1 CD34     1        380
##  8 358    CD34_1 CD34     1        572
##  9 359    CD34_1 CD34     1          0
## 10 360    CD34_1 CD34     1        320
## # … with 390 more rows
```

---
# Joining
Data frames can be joined on a shared variable _a.k.a._ a key. We want this key to be unique i.e. ENTREZ ID. 
---
## dplyr::inner_join - Merging dataframes

```r
tidy_counts
```

```
## # A tibble: 400 x 5
##    ENTREZ Sample CellType Rep   counts
##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;
##  1 350    CD34_1 CD34     1        204
##  2 351    CD34_1 CD34     1      15586
##  3 353    CD34_1 CD34     1        842
##  4 354    CD34_1 CD34     1          0
##  5 355    CD34_1 CD34     1        123
##  6 356    CD34_1 CD34     1          1
##  7 357    CD34_1 CD34     1        380
##  8 358    CD34_1 CD34     1        572
##  9 359    CD34_1 CD34     1          0
## 10 360    CD34_1 CD34     1        320
## # … with 390 more rows
```
---
## dplyr::inner_join - Merging dataframes

```r
counts_metadata
```

```
## # A tibble: 100 x 4
##    ID    SYMBOL  CHR   LENGTH
##    &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;int&gt;
##  1 350   APOH    chr17   1201
##  2 351   APP     chr21   4480
##  3 353   APRT    chr16    807
##  4 354   KLK3    chr19   1906
##  5 355   FAS     chr10   6691
##  6 356   FASLG   chr1    1859
##  7 357   SHROOM2 chrX    8206
##  8 358   AQP1    chr7    3786
##  9 359   AQP2    chr12   4179
## 10 360   AQP3    chr9    2950
## # … with 90 more rows
```
---
## dplyr::inner_join - Merging dataframes

```r
inner_join(tidy_counts, counts_metadata, by = c("ENTREZ" = "ID"))
```

```
## # A tibble: 400 x 8
##    ENTREZ Sample CellType Rep   counts SYMBOL  CHR   LENGTH
##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;int&gt;
##  1 350    CD34_1 CD34     1        204 APOH    chr17   1201
##  2 351    CD34_1 CD34     1      15586 APP     chr21   4480
##  3 353    CD34_1 CD34     1        842 APRT    chr16    807
##  4 354    CD34_1 CD34     1          0 KLK3    chr19   1906
##  5 355    CD34_1 CD34     1        123 FAS     chr10   6691
##  6 356    CD34_1 CD34     1          1 FASLG   chr1    1859
##  7 357    CD34_1 CD34     1        380 SHROOM2 chrX    8206
##  8 358    CD34_1 CD34     1        572 AQP1    chr7    3786
##  9 359    CD34_1 CD34     1          0 AQP2    chr12   4179
## 10 360    CD34_1 CD34     1        320 AQP3    chr9    2950
## # … with 390 more rows
```

---
## There are many ways to join things

Inner Join 

* Keeps all observations in x and y with matching keys

Outer Join

* A left join keeps all observations in x and those in y with matching keys.
* A right join keeps all observations in y and those in x with matching keys. 
* A full join keeps all observations in x and y

---
## EXAMPLE - Look at just expressed genes

```r
# In this pipe I group by gene, summarise the data based on the
# sum of counts, and filter for anything that has a count greater
# than 0. 
expressed_genes &lt;- tidy_counts %&gt;% 
  group_by(ENTREZ) %&gt;% 
  summarise(count_total=sum(counts)) %&gt;% 
  filter(count_total&gt;0)
expressed_genes
```

```
## # A tibble: 94 x 2
##    ENTREZ count_total
##    &lt;chr&gt;        &lt;int&gt;
##  1 350            307
##  2 351          26580
##  3 353           2471
##  4 355            569
##  5 356              2
##  6 357            560
##  7 358           7445
##  8 359             13
##  9 360           1982
## 10 361              1
## # … with 84 more rows
```
---
## dplyr::left_join 

```r
# Left join shows all genes as my full data frame tidy_counts is 
# used as the backbone. The filtered expressed genes is secondary, 
# and has missing values (unexpressed genes) which are filled with NA
left_join(tidy_counts, expressed_genes, by = c("ENTREZ" = "ENTREZ"))
```

```
## # A tibble: 400 x 6
##    ENTREZ Sample CellType Rep   counts count_total
##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;       &lt;int&gt;
##  1 350    CD34_1 CD34     1        204         307
##  2 351    CD34_1 CD34     1      15586       26580
##  3 353    CD34_1 CD34     1        842        2471
##  4 354    CD34_1 CD34     1          0          NA
##  5 355    CD34_1 CD34     1        123         569
##  6 356    CD34_1 CD34     1          1           2
##  7 357    CD34_1 CD34     1        380         560
##  8 358    CD34_1 CD34     1        572        7445
##  9 359    CD34_1 CD34     1          0          13
## 10 360    CD34_1 CD34     1        320        1982
## # … with 390 more rows
```
---
## dplyr::right_join

```r
# Right join shows only genes that survived filtering as it is using 
# the second dataframe as the backbone for the new dataframe. 
tidy_counts_expressed &lt;- right_join(tidy_counts, expressed_genes, by = c("ENTREZ" = "ENTREZ"))
tidy_counts_expressed %&gt;% print(n=20)
```

```
## # A tibble: 376 x 6
##    ENTREZ Sample  CellType Rep   counts count_total
##    &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;       &lt;int&gt;
##  1 350    CD34_1  CD34     1        204         307
##  2 350    ORTHO_1 ORTHO    1          0         307
##  3 350    CD34_2  CD34     2        103         307
##  4 350    ORTHO_2 ORTHO    2          0         307
##  5 351    CD34_1  CD34     1      15586       26580
##  6 351    ORTHO_1 ORTHO    1        479       26580
##  7 351    CD34_2  CD34     2      10476       26580
##  8 351    ORTHO_2 ORTHO    2         39       26580
##  9 353    CD34_1  CD34     1        842        2471
## 10 353    ORTHO_1 ORTHO    1        355        2471
## 11 353    CD34_2  CD34     2       1188        2471
## 12 353    ORTHO_2 ORTHO    2         86        2471
## 13 355    CD34_1  CD34     1        123         569
## 14 355    ORTHO_1 ORTHO    1        291         569
## 15 355    CD34_2  CD34     2        139         569
## 16 355    ORTHO_2 ORTHO    2         16         569
## 17 356    CD34_1  CD34     1          1           2
## 18 356    ORTHO_1 ORTHO    1          1           2
## 19 356    CD34_2  CD34     2          0           2
## 20 356    ORTHO_2 ORTHO    2          0           2
## # … with 356 more rows
```

---
## Filtering joins

Filtering joins 

* A semi join only keeps all observations in x that are matched in y. y isn't returned. 
* A anti join only keeps all observations in x that are _not_ matched in y. y isn't returned. 

---
## dplyr::semi_join

```r
# Semi join only keeps observations in x that are matched in y. y is 
# only used as a reference and is not in output
semi_join(tidy_counts, expressed_genes)
```

```
## Joining, by = "ENTREZ"
```

```
## # A tibble: 376 x 5
##    ENTREZ Sample CellType Rep   counts
##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;
##  1 350    CD34_1 CD34     1        204
##  2 351    CD34_1 CD34     1      15586
##  3 353    CD34_1 CD34     1        842
##  4 355    CD34_1 CD34     1        123
##  5 356    CD34_1 CD34     1          1
##  6 357    CD34_1 CD34     1        380
##  7 358    CD34_1 CD34     1        572
##  8 359    CD34_1 CD34     1          0
##  9 360    CD34_1 CD34     1        320
## 10 361    CD34_1 CD34     1          0
## # … with 366 more rows
```
---
## dplyr::anti_join

```r
# Anti join only keeps observations in x that are not matched in y. 
# y is only used as a reference and is not in output
anti_join(tidy_counts, expressed_genes)
```

```
## Joining, by = "ENTREZ"
```

```
## # A tibble: 24 x 5
##    ENTREZ Sample  CellType Rep   counts
##    &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;
##  1 354    CD34_1  CD34     1          0
##  2 398    CD34_1  CD34     1          0
##  3 401    CD34_1  CD34     1          0
##  4 417    CD34_1  CD34     1          0
##  5 474    CD34_1  CD34     1          0
##  6 495    CD34_1  CD34     1          0
##  7 354    ORTHO_1 ORTHO    1          0
##  8 398    ORTHO_1 ORTHO    1          0
##  9 401    ORTHO_1 ORTHO    1          0
## 10 417    ORTHO_1 ORTHO    1          0
## # … with 14 more rows
```

---
## CHALLENGE
 
* Calculate CPMs and TPMs

Hint:

**Counts per million (CPM)** are the gene counts normalized to total counts in a sample, multiplied by a million to give you a sensible number.  

gene_A_CPM = (gene_A_counts / sum(all_genes_counts)) * 1,000,000

**Transcripts per million (TPM)** are the gene counts normalized to total counts in a sample, multiplied by a million to give you a sensible number.  

gene_A_TPM = (gene_A_counts / sum(all_genes_counts / all_genes_lengths)) * 1/gene_A_length * 1,000,000

More info on RNAseq quantification here:
http://luisvalesilva.com/datasimple/rna-seq_units.html

---
## ANSWER

```r
# Use Group to focus computation on each sample. Use mutate to make 
# new variable that is the CPM
tidy_counts_expressed &lt;- tidy_counts_expressed  %&gt;% 
  group_by(Sample) %&gt;% 
  mutate(CPM=(counts/sum(counts))*1000000)

# Join our tidy data to the metadata, then make new variable that is the TPM
tidy_counts_expressed &lt;- tidy_counts_expressed %&gt;% 
  inner_join(counts_metadata, by = c("ENTREZ" = "ID")) %&gt;%  
  mutate(TPM=(counts/sum(counts/LENGTH))*(1000000/LENGTH))

# Simple X-Y plot comparing TPM and CPM. 
p &lt;- tidy_counts_expressed %&gt;% 
  ggplot(aes(x=CPM, y=TPM)) + 
  geom_point() + 
  scale_x_continuous(name="log2(CPM)",trans='log2') + 
  scale_y_continuous(name="log2(TPM)",trans='log2')
p
```

![](20191009_tidyr_files/figure-html/unnamed-chunk-68-1.png)&lt;!-- --&gt;

---
# Readr again: Writing out 
You've made a lovely new tibble file. Now you need to save it somewhere. 

```r
#Theres a wide range of writing  options. Can specify the delmiter directly or use a specific function
write_delim(tidy_counts_expressed, '../expressed_genes_output.csv', delim =',')

write_csv(tidy_counts_expressed, '../expressed_genes_output.csv')
```

A key difference compared to base is that it does not write out row names. Tibbles generally don't have rownames.

---
## tidyverse so far
__At this point we have covered or touched on the most essential facets of tidy__  
* ~~ggplot2 – making pretty graphs~~ 
* ~~readr – reading data into R~~  
* ~~dplyr – manipulating data~~  
* ~~tibble - working with tibbles~~  
* ~~tidyr – miscellaneous tools for tidying data~~
* purrr - iterating over data  
* stringr – working with strings  
* forcats - working with factors  

---
# purrr - Functional programming 

* Applying functions to datasets  
* Base people use _for_ loops or apply  
* Big advantage is that purrr readily handles nested dataframes and has standard outputs

---
## purrr::map - tidy way to iterate - (pt1)

```r
# Map is the tidy equivalent to apply. Here we take our untidy counts, 
# trim of IDs, and then calculate means for each column. By default the 
# output is a list
untidy_counts %&gt;% 
  dplyr::select(-ENTREZ) %&gt;% 
  map(mean)
```

```
## $CD34_1
## [1] 1497.67
## 
## $ORTHO_1
## [1] 822.33
## 
## $CD34_2
## [1] 1056.85
## 
## $ORTHO_2
## [1] 329.05
```
---
## purrr::map - tidy way to iterate - (pt2)

```r
# Same as the above line, but using map_dbl specifies the outputs is 
# going to be a double
untidy_counts %&gt;% 
  dplyr::select(-ENTREZ) %&gt;% 
  map_dbl(mean)
```

```
##  CD34_1 ORTHO_1  CD34_2 ORTHO_2 
## 1497.67  822.33 1056.85  329.05
```
---
## With tidy data - summarize sometimes

```r
# Summary sometimes also works in this context
tidy_counts %&gt;% 
  group_by(Sample) %&gt;% 
  summarize(mean_counts = mean(counts))
```

```
## # A tibble: 4 x 2
##   Sample  mean_counts
##   &lt;chr&gt;         &lt;dbl&gt;
## 1 CD34_1        1498.
## 2 CD34_2        1057.
## 3 ORTHO_1        822.
## 4 ORTHO_2        329.
```
---
## With tidy data - the purrr way

```r
# This is an alternative method for doing this with an tidied frame
tidy_counts %&gt;% 
  split(.$Sample) %&gt;% 
  map_dbl(~mean(.$counts))
```

```
##  CD34_1  CD34_2 ORTHO_1 ORTHO_2 
## 1497.67 1056.85  822.33  329.05
```
---
## purrr::pmap for multiple inputs

```r
# pmap is a map variant for dealing with multiple inputs. This can be 
# used to apply a function on a row by row basis
list(untidy_counts$ORTHO_1, untidy_counts$ORTHO_2) %&gt;% 
  pmap_dbl(mean)
```

```
##   [1]     0   479   355     0   291     1     3  2225    12   502     1     1     6     0     0     0     0  1034  1864   728     5  6697
##  [23]   314  4845     1   574  2265  1632 10977  2457   171   193     0   577   304   138   464    67   702   410     0    11   147     0
##  [45]   218    25   824   175     1   111  1819   197   217    45    15     0     1     0     5  1136   255 11889     0    13     0     0
##  [67]     1   151  2520     0   151    15     0     0   118   480   402  6132   771   628   783     0    83  3453     3   121     0     5
##  [89]    18  1392  1454     0    33  1118   153  1263     1     0  3585     0
```

---
## Nest - Values can be much more

Dataframes can be be simplified by making them more complex.  

Each value within a tibble can contian more abstract information then just numbers and chracters. Instead you can store another tibble, or object.

---
## tidyr::nest - (pt1)

```r
# Nest all the data by sample
tidy_counts_nest &lt;- tidy_counts_expressed %&gt;% 
  group_by(Sample) %&gt;%
  nest()

# Looking at tibble it is a new datatype that appears simplified
tidy_counts_nest
```

```
## # A tibble: 4 x 2
## # Groups:   Sample [4]
##   Sample             data
##   &lt;chr&gt;   &lt;list&lt;df[,10]&gt;&gt;
## 1 CD34_1        [94 × 10]
## 2 ORTHO_1       [94 × 10]
## 3 CD34_2        [94 × 10]
## 4 ORTHO_2       [94 × 10]
```

```r
tidy_counts_nest$data %&gt;% is()
```

```
## [1] "vctrs_list_of" "vctrs_vctr"    "oldClass"
```
---
## tidyr::nest - (pt2)

```r
# The data is still there, nested within one of the variables
tidy_counts_nest$data[[1]]
```

```
## # A tibble: 94 x 10
##    ENTREZ CellType Rep   counts count_total       CPM SYMBOL  CHR   LENGTH      TPM
##    &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;       &lt;int&gt;     &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;int&gt;    &lt;dbl&gt;
##  1 350    CD34     1        204         307   1362.   APOH    chr17   1201  3069.  
##  2 351    CD34     1      15586       26580 104068.   APP     chr21   4480 62851.  
##  3 353    CD34     1        842        2471   5622.   APRT    chr16    807 18849.  
##  4 355    CD34     1        123         569    821.   FAS     chr10   6691   332.  
##  5 356    CD34     1          1           2      6.68 FASLG   chr1    1859     9.72
##  6 357    CD34     1        380         560   2537.   SHROOM2 chrX    8206   837.  
##  7 358    CD34     1        572        7445   3819.   AQP1    chr7    3786  2729.  
##  8 359    CD34     1          0          13      0    AQP2    chr12   4179     0   
##  9 360    CD34     1        320        1982   2137.   AQP3    chr9    2950  1960.  
## 10 361    CD34     1          0           1      0    AQP4    chr18   5217     0   
## # … with 84 more rows
```
---
## tidyr::nest and purrr::map

```r
# Map can be used to apply functions across nested dataframes.
# Here we calculate a linear model. This is also saved in the tibble. 
tidy_counts_nest &lt;- tidy_counts_nest %&gt;% 
  mutate(my_model = map(data, ~lm(CPM ~ TPM, data = .)))

tidy_counts_nest
```

```
## # A tibble: 4 x 3
## # Groups:   Sample [4]
##   Sample             data my_model
##   &lt;chr&gt;   &lt;list&lt;df[,10]&gt;&gt; &lt;list&gt;  
## 1 CD34_1        [94 × 10] &lt;lm&gt;    
## 2 ORTHO_1       [94 × 10] &lt;lm&gt;    
## 3 CD34_2        [94 × 10] &lt;lm&gt;    
## 4 ORTHO_2       [94 × 10] &lt;lm&gt;
```

```r
tidy_counts_nest$my_model[[1]]
```

```
## 
## Call:
## lm(formula = CPM ~ TPM, data = .)
## 
## Coefficients:
## (Intercept)          TPM  
##   3864.9798       0.6367
```
---
## broom::tidy - making models legible 

```r
# Tidy also has the ability to "tidy" up outputs from common statistical 
# packages, using broom.
library(broom)

tidy_counts_nest &lt;- tidy_counts_nest %&gt;% 
  mutate(my_tidy_model = map(my_model, broom::tidy))

tidy_counts_nest
```

```
## # A tibble: 4 x 4
## # Groups:   Sample [4]
##   Sample             data my_model my_tidy_model   
##   &lt;chr&gt;   &lt;list&lt;df[,10]&gt;&gt; &lt;list&gt;   &lt;list&gt;          
## 1 CD34_1        [94 × 10] &lt;lm&gt;     &lt;tibble [2 × 5]&gt;
## 2 ORTHO_1       [94 × 10] &lt;lm&gt;     &lt;tibble [2 × 5]&gt;
## 3 CD34_2        [94 × 10] &lt;lm&gt;     &lt;tibble [2 × 5]&gt;
## 4 ORTHO_2       [94 × 10] &lt;lm&gt;     &lt;tibble [2 × 5]&gt;
```
---
## broom::tidy - making models legible 

```r
tidy_counts_nest$my_tidy_model[[1]]
```

```
## # A tibble: 2 x 5
##   term        estimate std.error statistic  p.value
##   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1 (Intercept) 3865.    1101.          3.51 6.94e- 4
## 2 TPM            0.637    0.0398     16.0  2.75e-28
```

---
## dplyr::unnest - expand out - (pt1)

```r
# Unnesting to get everything back into a dataframe is very straightforward
tidy_counts_nest %&gt;%
  unnest(my_tidy_model)
```

```
## # A tibble: 8 x 8
## # Groups:   Sample [4]
##   Sample             data my_model term        estimate std.error statistic  p.value
##   &lt;chr&gt;   &lt;list&lt;df[,10]&gt;&gt; &lt;list&gt;   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1 CD34_1        [94 × 10] &lt;lm&gt;     (Intercept) 3865.    1101.          3.51 6.94e- 4
## 2 CD34_1        [94 × 10] &lt;lm&gt;     TPM            0.637    0.0398     16.0  2.75e-28
## 3 ORTHO_1       [94 × 10] &lt;lm&gt;     (Intercept) 1522.     784.          1.94 5.53e- 2
## 4 ORTHO_1       [94 × 10] &lt;lm&gt;     TPM            0.857    0.0271     31.6  3.83e-51
## 5 CD34_2        [94 × 10] &lt;lm&gt;     (Intercept) 4083.    1085.          3.76 2.94e- 4
## 6 CD34_2        [94 × 10] &lt;lm&gt;     TPM            0.616    0.0378     16.3  7.03e-29
## 7 ORTHO_2       [94 × 10] &lt;lm&gt;     (Intercept) 1787.     937.          1.91 5.96e- 2
## 8 ORTHO_2       [94 × 10] &lt;lm&gt;     TPM            0.832    0.0333     25.0  8.45e-43
```
---
## dplyr::unnest - expand out - (pt2)

```r
# Unnesting can be done sequentially to keep adding to master dataframe
tidy_counts_nest %&gt;%
  unnest(my_tidy_model) %&gt;% 
  unnest(data)
```

```
## # A tibble: 752 x 17
## # Groups:   Sample [4]
##    Sample ENTREZ CellType Rep   counts count_total     CPM SYMBOL  CHR   LENGTH    TPM my_model term   estimate std.error statistic p.value
##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;       &lt;int&gt;   &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;int&gt;  &lt;dbl&gt; &lt;list&gt;   &lt;chr&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;
##  1 CD34_1 350    CD34     1        204         307  1.36e3 APOH    chr17   1201 3.07e3 &lt;lm&gt;     (Inte…    3865.     1101.      3.51 6.94e-4
##  2 CD34_1 351    CD34     1      15586       26580  1.04e5 APP     chr21   4480 6.29e4 &lt;lm&gt;     (Inte…    3865.     1101.      3.51 6.94e-4
##  3 CD34_1 353    CD34     1        842        2471  5.62e3 APRT    chr16    807 1.88e4 &lt;lm&gt;     (Inte…    3865.     1101.      3.51 6.94e-4
##  4 CD34_1 355    CD34     1        123         569  8.21e2 FAS     chr10   6691 3.32e2 &lt;lm&gt;     (Inte…    3865.     1101.      3.51 6.94e-4
##  5 CD34_1 356    CD34     1          1           2  6.68e0 FASLG   chr1    1859 9.72e0 &lt;lm&gt;     (Inte…    3865.     1101.      3.51 6.94e-4
##  6 CD34_1 357    CD34     1        380         560  2.54e3 SHROOM2 chrX    8206 8.37e2 &lt;lm&gt;     (Inte…    3865.     1101.      3.51 6.94e-4
##  7 CD34_1 358    CD34     1        572        7445  3.82e3 AQP1    chr7    3786 2.73e3 &lt;lm&gt;     (Inte…    3865.     1101.      3.51 6.94e-4
##  8 CD34_1 359    CD34     1          0          13  0.     AQP2    chr12   4179 0.     &lt;lm&gt;     (Inte…    3865.     1101.      3.51 6.94e-4
##  9 CD34_1 360    CD34     1        320        1982  2.14e3 AQP3    chr9    2950 1.96e3 &lt;lm&gt;     (Inte…    3865.     1101.      3.51 6.94e-4
## 10 CD34_1 361    CD34     1          0           1  0.     AQP4    chr18   5217 0.     &lt;lm&gt;     (Inte…    3865.     1101.      3.51 6.94e-4
## # … with 742 more rows
```

---
## CHALLENGE

1. Create a nested dataframe of df1. Specifically only of the two largest groups by common_name, and nest by the common_name. 2. Create a linear model between length_mm and IGF1_ng_ml using purrr over the nested dataframe. Add the results back to the dataframe
3. Create a new nested variable containing projections for length_mm based on your linear model (the function *predict* will be helpful here).

---
## ANSWER

```r
df1_nest &lt;- df1 %&gt;% 
  filter(common_name!='Coho salmon') %&gt;% 
  filter(common_name!='Sockeye salmon') %&gt;% 
  group_by(common_name) %&gt;% 
  nest()
df1_nest
```

```
## # A tibble: 2 x 2
## # Groups:   common_name [2]
##   common_name              data
##   &lt;chr&gt;          &lt;list&lt;df[,4]&gt;&gt;
## 1 Chinook salmon       [46 × 4]
## 2 Steelhead            [38 × 4]
```

```r
df1_nest &lt;- df1_nest %&gt;%
  mutate(my_model = map(data, ~lm(length_mm ~ IGF1_ng_ml, data = ., na.action = na.omit)))
df1_nest 
```

```
## # A tibble: 2 x 3
## # Groups:   common_name [2]
##   common_name              data my_model
##   &lt;chr&gt;          &lt;list&lt;df[,4]&gt;&gt; &lt;list&gt;  
## 1 Chinook salmon       [46 × 4] &lt;lm&gt;    
## 2 Steelhead            [38 × 4] &lt;lm&gt;
```

```r
df1_nest &lt;- df1_nest %&gt;%
  mutate(predictions=map(my_model, predict)) 
df1_nest 
```

```
## # A tibble: 2 x 4
## # Groups:   common_name [2]
##   common_name              data my_model predictions
##   &lt;chr&gt;          &lt;list&lt;df[,4]&gt;&gt; &lt;list&gt;   &lt;list&gt;     
## 1 Chinook salmon       [46 × 4] &lt;lm&gt;     &lt;dbl [42]&gt; 
## 2 Steelhead            [38 × 4] &lt;lm&gt;     &lt;dbl [35]&gt;
```

```r
df1_nest %&gt;% pull(predictions)
```

```
## [[1]]
##        1        2        3        4        5        6        7        8        9       10       11       12       13       14       15 
## 174.0853 204.0953 210.3557 231.1932 148.3513 259.7978 190.8976 171.3871 190.7028 191.2548 181.2579 122.7105 139.3573 147.1669 205.1610 
##       16       17       18       19       20       22       23       24       25       26       27       28       30       31       34 
## 178.0563 211.4999 232.4803 146.3425 176.3637 231.4895 232.0093 153.6838 244.9919 135.1426 181.6404 191.7068 202.8107 189.6438 219.2654 
##       35       36       37       38       39       40       41       42       43       44       45       46 
## 156.6304 180.8808 160.2568 177.5445 144.6843 131.2548 177.3148 176.9896 127.9072 196.2106 143.8471 178.5774 
## 
## [[2]]
##        1        3        5        6        7        8        9       10       11       12       13       14       15       16       17 
## 214.2574 213.1801 215.4682 215.3350 213.5525 214.6490 215.0501 215.0165 215.1124 212.9443 216.4120 216.0293 214.1118 214.5615 212.1104 
##       18       19       20       21       22       23       24       26       27       28       29       30       31       32       33 
## 214.9008 212.3449 213.9893 214.4727 214.2505 213.1433 212.0261 214.7772 215.5983 214.5010 215.3785 214.8688 215.1779 215.5824 212.6976 
##       34       35       36       37       38 
## 214.1116 214.7586 216.2151 215.3987 214.0160
```

---
# stringr - working with strings

If the data you are working with involves characters from data entry often there will be errors i.e. clinical study metadata or a hand-typed list of genes of interest. Tidying data also means fixing these problems. stringr helps make this easy. 

* Access and manipulate characters
* Deal with whitespace
* Pattern Recognition

Though stringr is pretty comprehensive and covers most of what you will need, there is a sister package called stringi with even more functionality. 

---
## stringr::strsub - string basics
Many overlapping functions with base for combining, subsetting, converting and finding strings


```r
brc &lt;- c("Tom", "Ji-Dung", "Matt")

# Extract substrings from a range. Here the 1st to 3rd character
brc %&gt;% str_sub(1, 3)
```

```
## [1] "Tom" "Ji-" "Mat"
```

```r
# Extract substrings from a range. Here the 2nd to 2nd to last character
brc %&gt;% str_sub(2, -2)
```

```
## [1] "o"     "i-Dun" "at"
```

```r
# Assign values back to substrings. Here the 2nd to 2nd to last character is replaced with X.
str_sub(brc, 2, -2) &lt;- 'X'
brc
```

```
## [1] "TXm" "JXg" "MXt"
```

---
## stringr::str_trim - stripping whitespace


```r
brc2 &lt;- c("Tom  ", "  Ji  -Dung", "Matt   ")

# Trim whitespace from strings
brc2 &lt;- str_trim(brc2)
brc2 
```

```
## [1] "Tom"       "Ji  -Dung" "Matt"
```

```r
# Can add whitespace to strings to get consistent length. Here all are 10 characters
str_pad(brc2, width=10, side='left')
```

```
## [1] "       Tom" " Ji  -Dung" "      Matt"
```

---
## stringr::str_to_* - Capitalization - (pt1)

```r
# Lets reuse our counts tibble. pull from dplyr can be used to grab a tibble 
# column and make it into a vector
tidy_counts_expressed %&gt;% 
  pull(SYMBOL) %&gt;%
  head()
```

```
## [1] "APOH" "APOH" "APOH" "APOH" "APP"  "APP"
```

```r
# Here we pull our gene symbols from our tibble into a vector, and then convert 
# them into title style capitalization
tidy_counts_expressed %&gt;% 
  pull(SYMBOL) %&gt;% 
  str_to_title() %&gt;% 
  head()
```

```
## [1] "Apoh" "Apoh" "Apoh" "Apoh" "App"  "App"
```
---
## stringr::str_to_* - Capitalization - (pt2)

```r
# String manipulation functions can be used on tibbles using mutate. Here we convert
# gene symbols to title style capitalization
tidy_counts_expressed %&gt;% 
  mutate(SYMBOL = str_to_title(SYMBOL))
```

```
## # A tibble: 376 x 11
## # Groups:   Sample [4]
##    ENTREZ Sample  CellType Rep   counts count_total     CPM SYMBOL CHR   LENGTH    TPM
##    &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;       &lt;int&gt;   &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt;  &lt;dbl&gt;
##  1 350    CD34_1  CD34     1        204         307   1362. Apoh   chr17   1201  3069.
##  2 350    ORTHO_1 ORTHO    1          0         307      0  Apoh   chr17   1201     0 
##  3 350    CD34_2  CD34     2        103         307    975. Apoh   chr17   1201  2027.
##  4 350    ORTHO_2 ORTHO    2          0         307      0  Apoh   chr17   1201     0 
##  5 351    CD34_1  CD34     1      15586       26580 104068. App    chr21   4480 62851.
##  6 351    ORTHO_1 ORTHO    1        479       26580   5825. App    chr21   4480  3333.
##  7 351    CD34_2  CD34     2      10476       26580  99125. App    chr21   4480 55281.
##  8 351    ORTHO_2 ORTHO    2         39       26580   1185. App    chr21   4480   703.
##  9 353    CD34_1  CD34     1        842        2471   5622. Aprt   chr16    807 18849.
## 10 353    ORTHO_1 ORTHO    1        355        2471   4317. Aprt   chr16    807 13711.
## # … with 366 more rows
```
---
## stringr::str_to_* - Capitalization - (pt3)

```r
# Here we convert chromosome annotation to capitals
tidy_counts_expressed %&gt;% 
  mutate(CHR = str_to_upper(CHR))
```

```
## # A tibble: 376 x 11
## # Groups:   Sample [4]
##    ENTREZ Sample  CellType Rep   counts count_total     CPM SYMBOL CHR   LENGTH    TPM
##    &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;       &lt;int&gt;   &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt;  &lt;dbl&gt;
##  1 350    CD34_1  CD34     1        204         307   1362. APOH   CHR17   1201  3069.
##  2 350    ORTHO_1 ORTHO    1          0         307      0  APOH   CHR17   1201     0 
##  3 350    CD34_2  CD34     2        103         307    975. APOH   CHR17   1201  2027.
##  4 350    ORTHO_2 ORTHO    2          0         307      0  APOH   CHR17   1201     0 
##  5 351    CD34_1  CD34     1      15586       26580 104068. APP    CHR21   4480 62851.
##  6 351    ORTHO_1 ORTHO    1        479       26580   5825. APP    CHR21   4480  3333.
##  7 351    CD34_2  CD34     2      10476       26580  99125. APP    CHR21   4480 55281.
##  8 351    ORTHO_2 ORTHO    2         39       26580   1185. APP    CHR21   4480   703.
##  9 353    CD34_1  CD34     1        842        2471   5622. APRT   CHR16    807 18849.
## 10 353    ORTHO_1 ORTHO    1        355        2471   4317. APRT   CHR16    807 13711.
## # … with 366 more rows
```

---
## stringr::detect - which have pattern?

```r
# Find patterns in different ways
# Detect gives a T/F whether the pattern 'salmon' is present in vector
df1 %&gt;% 
  pull(common_name) %&gt;% 
  str_detect('salmon')
```

```
##  [1]  TRUE  TRUE  TRUE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
## [23] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE
## [45]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE
## [67] FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
## [89]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
```
---
## stringr::detect - give me only patterns

```r
# Subset returns the match if the pattern 'salmon' is present in vector
df1 %&gt;% 
  dplyr::pull(common_name) %&gt;% 
  str_subset('salmon') 
```

```
##  [1] "Chinook salmon" "Sockeye salmon" "Sockeye salmon" "Chinook salmon" "Sockeye salmon" "Chinook salmon" "Chinook salmon"
##  [8] "Sockeye salmon" "Chinook salmon" "Chinook salmon" "Sockeye salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon"
## [15] "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon" "Coho salmon"    "Coho salmon"   
## [22] "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon"
## [29] "Sockeye salmon" "Sockeye salmon" "Sockeye salmon" "Sockeye salmon" "Sockeye salmon" "Chinook salmon" "Chinook salmon"
## [36] "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon"
## [43] "Chinook salmon" "Sockeye salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon"
## [50] "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon" "Chinook salmon"
## [57] "Chinook salmon" "Chinook salmon" "Chinook salmon"
```
---
## stringr::str_ends - ends with pattern?

```r
# Ends is similar to detect as it gives gives a T/F whether the pattern 'salmon' 
# is present in vector, but the pattern has to be at the end. 
df1 %&gt;% 
  dplyr::pull(common_name) %&gt;% 
  str_ends('salmon') 
```

```
##  [1]  TRUE  TRUE  TRUE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
## [23] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE
## [45]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE
## [67] FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
## [89]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
```

```r
df1 %&gt;% 
  filter(str_ends(common_name,'salmon'))
```

```
## # A tibble: 59 x 5
##    salmon_id common_name    age_classbylength  length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
##  1     35032 Chinook salmon yearling                 147       41.3
##  2     35035 Sockeye salmon juvenile                 121       NA  
##  3     35036 Sockeye salmon juvenile                 112       NA  
##  4     35033 Chinook salmon mixed age juvenile       444       62.1
##  5     35034 Sockeye salmon juvenile                 139       NA  
##  6     35142 Chinook salmon yearling                 149       66.5
##  7     35143 Chinook salmon yearling                 204       80.9
##  8     35144 Sockeye salmon juvenile                 140       NA  
##  9     35145 Chinook salmon yearling                 130       23.4
## 10     35146 Chinook salmon mixed age juvenile       422      101. 
## # … with 49 more rows
```
---
## stringr::str_count - how many patterns?

```r
#Count gives you the total number of times your pattern appears in each chracter in the vector
df1 %&gt;% 
  dplyr::pull(common_name) %&gt;% 
  str_count('salmon')
```

```
##  [1] 1 1 1 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0
## [68] 0 1 1 1 1 1 1 1 1 1 1 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
```

```r
df1 %&gt;% 
  dplyr::pull(common_name) %&gt;% 
  str_count('o')
```

```
##  [1] 3 2 2 0 0 3 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 3 3 2 3 3 2 0 0 0 0 0 0 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 2 2 2 2 2 0 3 0
## [68] 0 3 3 3 3 3 3 3 3 3 2 0 0 0 0 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3
```
---
## stringr::str_replace_all - replace a string

```r
#Replace
df1 %&gt;% 
  dplyr::pull(common_name) %&gt;%
  str_replace_all('Steelhead','Steelhead trout' )
```

```
##  [1] "Chinook salmon"  "Sockeye salmon"  "Sockeye salmon"  "Steelhead trout" "Steelhead trout" "Chinook salmon"  "Sockeye salmon" 
##  [8] "Steelhead trout" "Steelhead trout" "Steelhead trout" "Steelhead trout" "Steelhead trout" "Steelhead trout" "Steelhead trout"
## [15] "Steelhead trout" "Steelhead trout" "Steelhead trout" "Steelhead trout" "Steelhead trout" "Steelhead trout" "Steelhead trout"
## [22] "Steelhead trout" "Steelhead trout" "Steelhead trout" "Steelhead trout" "Steelhead trout" "Steelhead trout" "Steelhead trout"
## [29] "Steelhead trout" "Steelhead trout" "Chinook salmon"  "Chinook salmon"  "Sockeye salmon"  "Chinook salmon"  "Chinook salmon" 
## [36] "Sockeye salmon"  "Steelhead trout" "Steelhead trout" "Steelhead trout" "Steelhead trout" "Steelhead trout" "Steelhead trout"
## [43] "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Chinook salmon" 
## [50] "Chinook salmon"  "Coho salmon"     "Coho salmon"     "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Chinook salmon" 
## [57] "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Sockeye salmon"  "Sockeye salmon"  "Sockeye salmon"  "Sockeye salmon" 
## [64] "Sockeye salmon"  "Steelhead trout" "Chinook salmon"  "Steelhead trout" "Steelhead trout" "Chinook salmon"  "Chinook salmon" 
## [71] "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Chinook salmon" 
## [78] "Sockeye salmon"  "Steelhead trout" "Steelhead trout" "Steelhead trout" "Steelhead trout" "Chinook salmon"  "Chinook salmon" 
## [85] "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Chinook salmon" 
## [92] "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Chinook salmon"  "Chinook salmon"
```
---
## stringr::str_replace_all - replace a string

```r
df1 %&gt;% 
  mutate(common_name = str_replace_all(common_name,'Steelhead','Steelhead trout' ))
```

```
## # A tibble: 97 x 5
##    salmon_id common_name     age_classbylength  length_mm IGF1_ng_ml
##        &lt;dbl&gt; &lt;chr&gt;           &lt;chr&gt;                  &lt;dbl&gt;      &lt;dbl&gt;
##  1     35032 Chinook salmon  yearling                 147       41.3
##  2     35035 Sockeye salmon  juvenile                 121       NA  
##  3     35036 Sockeye salmon  juvenile                 112       NA  
##  4     35037 Steelhead trout juvenile                 220       42.7
##  5     35038 Steelhead trout juvenile                 152       NA  
##  6     35033 Chinook salmon  mixed age juvenile       444       62.1
##  7     35034 Sockeye salmon  juvenile                 139       NA  
##  8     35048 Steelhead trout juvenile                 288       24.2
##  9     35049 Steelhead trout juvenile                 190       NA  
## 10     35050 Steelhead trout juvenile                 283       63.5
## # … with 87 more rows
```

---
## CHALLENGE

1. Are there any Hox genes (starting with HOX), or ATPase genes (starting with ATP), in our expressed genes list?
2. Subset to just ATPase genes. Create a new lowercase variable that has 'ATP' removed from the symbol.

---
## ANSWER

```r
tidy_counts_expressed %&gt;% 
  mutate(ATPase = str_starts(SYMBOL,'ATP')) %&gt;% 
  split(.$Sample) %&gt;% 
  map(~sum(.$ATPase))
```

```
## $CD34_1
## [1] 15
## 
## $CD34_2
## [1] 15
## 
## $ORTHO_1
## [1] 15
## 
## $ORTHO_2
## [1] 15
```

```r
tidy_counts_expressed %&gt;% 
  mutate(Hox = str_starts(SYMBOL,'HOX')) %&gt;%
  split(.$Sample) %&gt;% 
  map(~sum(.$HOX))
```

```
## Warning: Unknown or uninitialised column: 'HOX'.

## Warning: Unknown or uninitialised column: 'HOX'.

## Warning: Unknown or uninitialised column: 'HOX'.

## Warning: Unknown or uninitialised column: 'HOX'.
```

```
## $CD34_1
## [1] 0
## 
## $CD34_2
## [1] 0
## 
## $ORTHO_1
## [1] 0
## 
## $ORTHO_2
## [1] 0
```

```r
tidy_counts_expressed %&gt;% 
  filter(str_starts(SYMBOL,'ATP')) %&gt;%
  mutate(ATPtype = str_replace_all(SYMBOL,'ATP','' )) %&gt;%
  mutate(ATPtype = str_to_lower(ATPtype))
```

```
## # A tibble: 60 x 12
## # Groups:   Sample [4]
##    ENTREZ Sample  CellType Rep   counts count_total     CPM SYMBOL CHR   LENGTH     TPM ATPtype
##    &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;       &lt;int&gt;   &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt;   &lt;dbl&gt; &lt;chr&gt;  
##  1 476    CD34_1  CD34     1       4952       14023 33065.  ATP1A1 chr1    5912 15132.  1a1    
##  2 476    ORTHO_1 ORTHO    1       3453       14023 41990.  ATP1A1 chr1    5912 18205.  1a1    
##  3 476    CD34_2  CD34     2       4202       14023 39760.  ATP1A1 chr1    5912 16803.  1a1    
##  4 476    ORTHO_2 ORTHO    2       1416       14023 43033.  ATP1A1 chr1    5912 19342.  1a1    
##  5 477    CD34_1  CD34     1         13          44    86.8 ATP1A2 chr1    5972    39.3 1a2    
##  6 477    ORTHO_1 ORTHO    1          3          44    36.5 ATP1A2 chr1    5972    15.7 1a2    
##  7 477    CD34_2  CD34     2         26          44   246.  ATP1A2 chr1    5972   103.  1a2    
##  8 477    ORTHO_2 ORTHO    2          2          44    60.8 ATP1A2 chr1    5972    27.0 1a2    
##  9 478    CD34_1  CD34     1         78         266   521.  ATP1A3 chr19   4054   348.  1a3    
## 10 478    ORTHO_1 ORTHO    1        121         266  1471.  ATP1A3 chr19   4054   930.  1a3    
## # … with 50 more rows
```

---
# forcats - Handling factors

Factors are a data type that R uses to handle fixed categorical variables that have a known set of possible values. 

Factors are ordered, allowing hierachy to be preserved in relatively simple vectors.

---
## Making a factor - Vectors
_[This is base]_

```r
# Vectors are easy to turn into factors with factor()
tidy_counts_expressed_samples &lt;- tidy_counts_expressed %&gt;% 
  pull(Sample) %&gt;% 
  factor() 

tidy_counts_expressed_samples %&gt;% head(n=10)
```

```
##  [1] CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1
## Levels: CD34_1 CD34_2 ORTHO_1 ORTHO_2
```
---
## Making a factor - Tibbles

```r
# Can also modify the data type of a tibble column with as_facotr, in an approach we have used before.
tidy_counts_expressed %&gt;% 
  ungroup() %&gt;% 
  mutate_at(vars(Sample), as_factor) 
```

```
## # A tibble: 376 x 11
##    ENTREZ Sample  CellType Rep   counts count_total     CPM SYMBOL CHR   LENGTH    TPM
##    &lt;chr&gt;  &lt;fct&gt;   &lt;chr&gt;    &lt;chr&gt;  &lt;int&gt;       &lt;int&gt;   &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt;  &lt;dbl&gt;
##  1 350    CD34_1  CD34     1        204         307   1362. APOH   chr17   1201  3069.
##  2 350    ORTHO_1 ORTHO    1          0         307      0  APOH   chr17   1201     0 
##  3 350    CD34_2  CD34     2        103         307    975. APOH   chr17   1201  2027.
##  4 350    ORTHO_2 ORTHO    2          0         307      0  APOH   chr17   1201     0 
##  5 351    CD34_1  CD34     1      15586       26580 104068. APP    chr21   4480 62851.
##  6 351    ORTHO_1 ORTHO    1        479       26580   5825. APP    chr21   4480  3333.
##  7 351    CD34_2  CD34     2      10476       26580  99125. APP    chr21   4480 55281.
##  8 351    ORTHO_2 ORTHO    2         39       26580   1185. APP    chr21   4480   703.
##  9 353    CD34_1  CD34     1        842        2471   5622. APRT   chr16    807 18849.
## 10 353    ORTHO_1 ORTHO    1        355        2471   4317. APRT   chr16    807 13711.
## # … with 366 more rows
```
---
## Making a factor - Tibbles

```r
tidy_counts_expressed %&gt;% 
  ungroup() %&gt;% 
  mutate_at(vars(Sample), as_factor) %&gt;% 
  pull(Sample) %&gt;% head(n=10)
```

```
##  [1] CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1
## Levels: CD34_1 ORTHO_1 CD34_2 ORTHO_2
```
---
## Making a factor - Controlling levels

```r
# When you factorize you can use a vector to determine the order
my_levels1&lt;-c('ORTHO_1','ORTHO_2','CD34_1','CD34_2')
tidy_counts_expressed %&gt;% 
  pull(Sample) %&gt;% 
  factor(levels = my_levels1 ) %&gt;% head(n=10)
```

```
##  [1] CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1
## Levels: ORTHO_1 ORTHO_2 CD34_1 CD34_2
```

```r
# When you factorize anything not in the given levels is turned to NA
my_levels2&lt;-c('ORTHO_1','CD34_1')
tidy_counts_expressed %&gt;% 
  pull(Sample) %&gt;% 
  factor(levels = my_levels2 ) %&gt;% head(n=10)
```

```
##  [1] CD34_1  ORTHO_1 &lt;NA&gt;    &lt;NA&gt;    CD34_1  ORTHO_1 &lt;NA&gt;    &lt;NA&gt;    CD34_1  ORTHO_1
## Levels: ORTHO_1 CD34_1
```

```r
# Its straightforward to grab the levels from the factor
tidy_counts_expressed_samples %&gt;% 
  levels()
```

```
## [1] "CD34_1"  "CD34_2"  "ORTHO_1" "ORTHO_2"
```

---
## Why do we factorize? For order. 

```r
p &lt;- tidy_counts_expressed %&gt;% 
  ungroup() %&gt;% 
  mutate_at(vars(Sample), as_factor) %&gt;% 
  group_by(Sample) %&gt;% 
  summarize(mean_count=mean(counts)) %&gt;% 
  ggplot(aes(x=Sample, y= mean_count)) + 
  geom_bar(stat = "identity") + 
  theme(axis.text.x = element_text(angle = 90))
```
---
## Why do we factorize? For order. 

```r
p
```

![](20191009_tidyr_files/figure-html/unnamed-chunk-100-1.png)&lt;!-- --&gt;

---
## forcats::fct_relevel - Reorder manually

```r
# fct_relevel - reorder manually
tidy_counts_expressed %&gt;% 
  ungroup() %&gt;% 
  mutate_at(vars(Sample), as_factor) %&gt;%
  mutate(Sample = fct_relevel(Sample, my_levels1)) %&gt;% 
  pull(Sample) %&gt;% head(n=10)
```

```
##  [1] CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1
## Levels: ORTHO_1 ORTHO_2 CD34_1 CD34_2
```

```r
# fct_relevel - reorder manually
tidy_counts_expressed %&gt;% 
  ungroup() %&gt;% 
  mutate_at(vars(Sample), as_factor) %&gt;%
  mutate(Sample = fct_relevel(Sample, my_levels2)) %&gt;% 
  pull(Sample) %&gt;% head(n=10)
```

```
##  [1] CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1
## Levels: ORTHO_1 CD34_1 CD34_2 ORTHO_2
```
---
## forcats::fct_reorder - Reorder using data

```r
# fct_reorder - reorder based on the data. Here we are ordering based
# on mean counts for each sample.
tidy_counts_expressed %&gt;%
  ungroup() %&gt;% 
  mutate_at(vars(Sample), as_factor) %&gt;% 
  mutate(Sample = fct_reorder(Sample, counts, mean)) %&gt;% 
  pull(Sample) %&gt;% head(n=10)
```

```
##  [1] CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1 CD34_2  ORTHO_2 CD34_1  ORTHO_1
## Levels: ORTHO_2 ORTHO_1 CD34_2 CD34_1
```

---
## Now our ordering can be fixed

```r
p &lt;- tidy_counts_expressed %&gt;% 
  ungroup() %&gt;% 
  mutate_at(vars(Sample), as_factor) %&gt;%
  mutate(Sample=fct_relevel(Sample, my_levels1)) %&gt;% 
  group_by(Sample) %&gt;% 
  summarize(mean_count=mean(counts)) %&gt;% 
  ggplot(aes(x=Sample, y= mean_count)) + 
  geom_bar(stat = "identity") + 
  theme(axis.text.x = element_text(angle = 90))
```
---
## Now our ordering can be fixed 

```r
p
```

![](20191009_tidyr_files/figure-html/unnamed-chunk-104-1.png)&lt;!-- --&gt;

---
## forcats::fct_recode - Change levels

```r
df1$age_classbylength
```

```
##  [1] "yearling"           "juvenile"           "juvenile"           "juvenile"           "juvenile"           "mixed age juvenile"
##  [7] "juvenile"           "juvenile"           "juvenile"           "juvenile"           "juvenile"           "juvenile"          
## [13] "juvenile"           "juvenile"           "juvenile"           "juvenile"           "juvenile"           "juvenile"          
## [19] "juvenile"           "juvenile"           "juvenile"           "juvenile"           "juvenile"           "juvenile"          
## [25] "juvenile"           "juvenile"           "juvenile"           "juvenile"           "juvenile"           "juvenile"          
## [31] "yearling"           "yearling"           "juvenile"           "yearling"           "mixed age juvenile" "juvenile"          
## [37] "juvenile"           "juvenile"           "juvenile"           "juvenile"           "juvenile"           "juvenile"          
## [43] "yearling"           "yearling"           "yearling"           "yearling"           "yearling"           "yearling"          
## [49] "yearling"           "yearling"           "yearling"           "yearling"           "yearling"           "yearling"          
## [55] "yearling"           "yearling"           "yearling"           "yearling"           "subyearling"        "juvenile"          
## [61] "juvenile"           "juvenile"           "juvenile"           "juvenile"           "juvenile"           "yearling"          
## [67] "juvenile"           "juvenile"           "mixed age juvenile" "yearling"           "yearling"           "yearling"          
## [73] "yearling"           "yearling"           "yearling"           "yearling"           "yearling"           "juvenile"          
## [79] "juvenile"           "juvenile"           "juvenile"           "juvenile"           "yearling"           "yearling"          
## [85] "yearling"           "yearling"           "yearling"           "yearling"           "yearling"           "yearling"          
## [91] "yearling"           "yearling"           "yearling"           "yearling"           "yearling"           "yearling"          
## [97] "yearling"
```

```r
# Recoding levels to give them a new name
df1 %&gt;% 
  mutate_at(vars(age_classbylength), as_factor) %&gt;%
  mutate(age_classbylength=fct_recode(age_classbylength, "mixed juvenile" = "mixed age juvenile")) %&gt;%
  pull(age_classbylength)
```

```
##  [1] yearling       juvenile       juvenile       juvenile       juvenile       mixed juvenile juvenile       juvenile       juvenile      
## [10] juvenile       juvenile       juvenile       juvenile       juvenile       juvenile       juvenile       juvenile       juvenile      
## [19] juvenile       juvenile       juvenile       juvenile       juvenile       juvenile       juvenile       juvenile       juvenile      
## [28] juvenile       juvenile       juvenile       yearling       yearling       juvenile       yearling       mixed juvenile juvenile      
## [37] juvenile       juvenile       juvenile       juvenile       juvenile       juvenile       yearling       yearling       yearling      
## [46] yearling       yearling       yearling       yearling       yearling       yearling       yearling       yearling       yearling      
## [55] yearling       yearling       yearling       yearling       subyearling    juvenile       juvenile       juvenile       juvenile      
## [64] juvenile       juvenile       yearling       juvenile       juvenile       mixed juvenile yearling       yearling       yearling      
## [73] yearling       yearling       yearling       yearling       yearling       juvenile       juvenile       juvenile       juvenile      
## [82] juvenile       yearling       yearling       yearling       yearling       yearling       yearling       yearling       yearling      
## [91] yearling       yearling       yearling       yearling       yearling       yearling       yearling      
## Levels: yearling juvenile mixed juvenile subyearling
```

```r
# Recoding levels to change to add one factor to another factor
df1 %&gt;% 
  mutate_at(vars(age_classbylength), as_factor) %&gt;% 
  mutate(age_classbylength=fct_recode(age_classbylength, "juvenile" = "mixed age juvenile")) %&gt;%
  pull(age_classbylength)
```

```
##  [1] yearling    juvenile    juvenile    juvenile    juvenile    juvenile    juvenile    juvenile    juvenile    juvenile    juvenile   
## [12] juvenile    juvenile    juvenile    juvenile    juvenile    juvenile    juvenile    juvenile    juvenile    juvenile    juvenile   
## [23] juvenile    juvenile    juvenile    juvenile    juvenile    juvenile    juvenile    juvenile    yearling    yearling    juvenile   
## [34] yearling    juvenile    juvenile    juvenile    juvenile    juvenile    juvenile    juvenile    juvenile    yearling    yearling   
## [45] yearling    yearling    yearling    yearling    yearling    yearling    yearling    yearling    yearling    yearling    yearling   
## [56] yearling    yearling    yearling    subyearling juvenile    juvenile    juvenile    juvenile    juvenile    juvenile    yearling   
## [67] juvenile    juvenile    juvenile    yearling    yearling    yearling    yearling    yearling    yearling    yearling    yearling   
## [78] juvenile    juvenile    juvenile    juvenile    juvenile    yearling    yearling    yearling    yearling    yearling    yearling   
## [89] yearling    yearling    yearling    yearling    yearling    yearling    yearling    yearling    yearling   
## Levels: yearling juvenile subyearling
```
---
## forcats::fct_count() - Summarise by counting

```r
# fct_count - Count up the number of each
df1 %&gt;% 
  mutate_at(vars(age_classbylength), as_factor) %&gt;% 
  pull(age_classbylength) %&gt;% 
  fct_count()
```

```
## # A tibble: 4 x 2
##   f                      n
##   &lt;fct&gt;              &lt;int&gt;
## 1 yearling              44
## 2 juvenile              49
## 3 mixed age juvenile     3
## 4 subyearling            1
```
---
## forcats::fct_infreq() - Mask rare factors

```r
# fct_infreq - mask rare factors by giving them a general summary term i.e. Other
df1 %&gt;% 
  mutate_at(vars(age_classbylength), as_factor) %&gt;%
  mutate(age_classbylength=fct_lump(age_classbylength, n=2)) %&gt;%
  pull(age_classbylength)
```

```
##  [1] yearling juvenile juvenile juvenile juvenile Other    juvenile juvenile juvenile juvenile juvenile juvenile juvenile juvenile juvenile
## [16] juvenile juvenile juvenile juvenile juvenile juvenile juvenile juvenile juvenile juvenile juvenile juvenile juvenile juvenile juvenile
## [31] yearling yearling juvenile yearling Other    juvenile juvenile juvenile juvenile juvenile juvenile juvenile yearling yearling yearling
## [46] yearling yearling yearling yearling yearling yearling yearling yearling yearling yearling yearling yearling yearling Other    juvenile
## [61] juvenile juvenile juvenile juvenile juvenile yearling juvenile juvenile Other    yearling yearling yearling yearling yearling yearling
## [76] yearling yearling juvenile juvenile juvenile juvenile juvenile yearling yearling yearling yearling yearling yearling yearling yearling
## [91] yearling yearling yearling yearling yearling yearling yearling
## Levels: yearling juvenile Other
```

```r
df1 %&gt;% 
  mutate_at(vars(age_classbylength), as_factor) %&gt;%
  mutate(age_classbylength=fct_lump(age_classbylength, n=2)) %&gt;%
  pull(age_classbylength) %&gt;% 
  fct_count()
```

```
## # A tibble: 3 x 2
##   f            n
##   &lt;fct&gt;    &lt;int&gt;
## 1 yearling    44
## 2 juvenile    49
## 3 Other        4
```
---
## forcats::fct_c - Combining factors

```r
# Normally facotrs do not like to be combined as levels in one may not exist in the other.
# Factor concatenation with fct_c help get around this. 
A &lt;- factor(c('Tom','Ji-Dung'))
B &lt;- factor('Matt')
fct_c(A, B)
```

```
## [1] Tom     Ji-Dung Matt   
## Levels: Ji-Dung Tom Matt
```

---
## CHALLENGE
1. Create new categorical variable based on whether a gene is short (0-3000), medium (0-6000) or long (6000+)
2. Ensure categorical variable is a factor which is ordered from long, medium then short
3. Draw a boxplot of TPMs between our new groups

---
## ANSWER

```r
tidy_counts_expressed %&gt;%
  mutate(length_cat=if_else(LENGTH&lt;3000, 'short', if_else(LENGTH&gt;6000, 'long', 'medium'))) %&gt;%
  mutate(length_cat=as_factor(length_cat)) %&gt;%
  mutate(length_cat=fct_relevel(length_cat,levels=c('long', 'medium','short')))  %&gt;% ggplot(aes(x=length_cat, y=TPM)) + geom_boxplot() + scale_y_continuous(trans='log2')
```

```
## Warning: Outer names are only allowed for unnamed scalar atomic inputs

## Warning: Outer names are only allowed for unnamed scalar atomic inputs

## Warning: Outer names are only allowed for unnamed scalar atomic inputs

## Warning: Outer names are only allowed for unnamed scalar atomic inputs
```

```
## Warning: Transformation introduced infinite values in continuous y-axis
```

```
## Warning: Removed 46 rows containing non-finite values (stat_boxplot).
```

![](20191009_tidyr_files/figure-html/unnamed-chunk-109-1.png)&lt;!-- --&gt;
---
# Summary 
We have covered the core packages in the tidyverse  

* ~~ggplot2 – making pretty graphs~~ 
* ~~readr – reading data into R~~  
* ~~dplyr – manipulating data~~  
* ~~tibble - working with tibbles~~  
* ~~tidyr – miscellaneous tools for tidying data~~
* ~~purrr - iterating over data~~
* ~~stringr – working with strings~~  
* ~~forcats - working with factors~~  

---
## Tidy beyond this workshop

Hadley Wickham (Chief Scientist at RStudio) is the driving force behind the tidyverse.  

Hadley wrote a paper about why he thinks tidy data is best: www.jstatsoft.org/v59/i10/paper.  

There is a lot of support for all things tidy at: https://www.tidyverse.org/  
(This includes really great cheat sheets for each tool)

---
## Tidy packages to check out:

plyranges: dplyr equivalent for working with ranges i.e. Granges

readxl: This package is very useful when you want to import Excel sheets in R

googledrive: Interact with your googledrive through R

lubridate and hms: Allow managin of calendar and time formats

magrittr: piping tools

broom: helps tidy up standard base function i.e. lm or t.test

tidymodels: A collection of tools for preparing for and validating model functions

---
## Other Good Resources

ggplot  
https://rockefelleruniversity.github.io/Plotting_In_R/

R for Data Science text book  
https://r4ds.had.co.nz/

Data Science with R text book
http://garrettgman.github.io/

Cancer Research UK - Intermediate R course
https://bioinformatics-core-shared-training.github.io/r-intermediate/

---
# Thanks

_BRC_
Tom Carroll: tcarroll@rockefeller.edu
Ji-Dung Luo: jluo@rockefeller.edu
Matt Paul: mpaul@rockefeller.edu

_Data Science Group_
Eli Stoyanova
Doug Barrows
Greg Gedman
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
